

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Module 4 - Investigating gene flow at loci of interest using haplotype networks &#8212; Training course in data analysis for genomic surveillance of African malaria vectors</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/myfile.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="//cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'workshop-7/module-4-haplotype-networks';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Workshop 8 - Analysis planning and reporting" href="../workshop-8/about.html" />
    <link rel="prev" title="Module 3 - Investigating gene flow at loci of interest using haplotype clustering" href="module-3-haplotype-clustering.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../home.html">
  
  
  
  
  
  
    <p class="title logo__title">Training course in data analysis for genomic surveillance of African malaria vectors</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../home.html">
                    Home
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 1</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-1/about.html">Workshop 1 - Pyrethroid target-site resistance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-1/module-1-colab.html">Module 1 - Interactive cloud computing with Google Colaboratory</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-1/module-2-sample-metadata.html">Module 2 - Accessing and exploring <em>Anopheles</em> genomic data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-1/module-3-genome.html">Module 3 - Navigating the <em>Anopheles gambiae</em> reference genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-1/module-4-vgsc-snps.html">Module 4 - Analysing SNPs in the <em>Vgsc</em> gene</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-1/journal-club.html">Journal club - Clarkson et al. (2021)</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 2</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-2/about.html">Workshop 2 - Metabolic insecticide resistance and copy number variation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-2/module-1-pandas.html">Module 1 - Introduction to pandas DataFrames</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-2/module-2-metabolic-ir.html">Module 2 - The genetic basis of metabolic resistance to insecticides in <em>Anopheles</em> mosquitoes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-2/module-3-cnv-calling.html">Module 3 - Detecting CNVs from genome sequence data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-2/module-4-cnv-frequencies.html">Module 4 - Analysing CNV frequencies at metabolic resistance genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-2/journal-club.html">Journal club - Lucas et al. (2019)</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 3</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-3/about.html">Workshop 3 - Population structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-3/module-1-plotly.html">Module 1 - Plotting with Plotly Express</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-3/module-2-population-structure.html">Module 2 - The biology of population structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-3/module-3-snps.html">Module 3 - Genome-wide SNP data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-3/module-4-NJT.html">Module 4 - Detecting taxonomic and geographical population structure using neighbour-joining trees (NJT)</a></li>






<li class="toctree-l1"><a class="reference internal" href="../workshop-3/journal-club.html">Journal club - Ag1000G Consortium (2017)</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 4</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-4/about.html">Workshop 4 - Cryptic species</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-4/module-1-numpy.html">Module 1 - NumPy arrays</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-4/module-2-cryptic-species.html">Module 2 - Species complexes and cryptic taxa in malaria vectors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-4/module-3-aims.html">Module 3 - Ancestry-informative markers (AIMs)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-4/module-4-taxon-analysis.html">Module 4 - Discovering cryptic taxa</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-4/journal-club.html">Journal club - Tennessen et al. (2020)</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 5</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-5/about.html">Workshop 5 - Genetic diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-5/module-1-xarray.html">Module 1 - Xarray datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-5/module-2-diversity.html">Module 2 - Introduction to genetic diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-5/module-3-stats.html">Module 3 - Genetic diversity summary statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-5/module-4-roh.html">Module 4 - Heterozygosity and runs of homozygosity</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 6</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-6/about.html">Workshop 6 - Detecting genes under recent positive selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-6/module-1-haplotypes.html">Module 1 - Haplotypes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-6/module-2-selection.html">Module 2 - The biology of selective sweeps</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-6/module-3-gwss.html">Module 3 - Detecting new forms of insecticide resistance using selection scans</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-6/module-4-primers.html">Module 4 - AnoPrimer - Primer Design in <em>Anopheles gambiae</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 7</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="about.html">Workshop 7 - Gene flow and the spread of insecticide resistance</a></li>
<li class="toctree-l1"><a class="reference internal" href="module-1-op-ir.html">Module 1 - Organophosphate resistance markers (combining SNP and CNV data)</a></li>
<li class="toctree-l1"><a class="reference internal" href="module-2-gene-flow.html">Module 2 - Adaptive gene flow</a></li>
<li class="toctree-l1"><a class="reference internal" href="module-3-haplotype-clustering.html">Module 3 - Investigating gene flow at loci of interest using haplotype clustering</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Module 4 - Investigating gene flow at loci of interest using haplotype networks</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 8</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-8/about.html">Workshop 8 - Analysis planning and reporting</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-8/module-1-planning-analysis.html">Module 1 - Planning a genomic surveillance exploratory data analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-8/module-2-presenting-results.html">Module 2 - Presenting results of a genomic surveillance data analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-8/module-3-genomic-surveillance-tanzania.html">Module 3 - Genomic surveillance of malaria vectors in Tanzania</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-8/module-4-genomic-surveillance-ghana.html">Module 4 - Genomic surveillance of malaria vectors in Ghana</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Malaria Vector Genome Observatory</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../general/data-access.html">Vector Observatory Data Access</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://colab.research.google.com/github/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io/blob/master/docs/workshop-7/module-4-haplotype-networks.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch onColab"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img src="../_static/images/logo_colab.png">
  </span>
<span class="btn__text-container">Colab</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io/issues/new?title=Issue%20on%20page%20%2Fworkshop-7/module-4-haplotype-networks.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/workshop-7/module-4-haplotype-networks.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Module 4 - Investigating gene flow at loci of interest using haplotype networks</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-objectives">Learning objectives</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#lecture">Lecture</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#english">English</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#francais">Français</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup">Setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#haplotype-networks">Haplotype networks</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#technical-interpretation">Technical interpretation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evolutionary-interpretation">Evolutionary interpretation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-1">Exercise 1</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">English</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Français</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-ace1">Example: <em>Ace1</em></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ace1-hierarchical-clustering-re-cap"><em>Ace1</em> hierarchical clustering re-cap</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ace1-haplotype-networks"><em>Ace1</em> haplotype networks</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-2">Exercise 2</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">English</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Français</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-gste2">Example: <em>Gste2</em></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-3">Exercise 3</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">English</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id6">Français</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#well-done">Well done!</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exercises">Exercises</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id7">English</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id8">Français</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <p><img alt="banner" src="https://anopheles-genomic-surveillance.github.io/_images/banner.jpg" /></p>
<p><em><strong>Training course in data analysis for genomic surveillance of African malaria vectors</strong></em></p>
<p><strong><a class="reference internal" href="about.html"><span class="doc std std-doc">Workshop 7</span></a> - Gene flow and the spread of insecticide resistance</strong></p>
<hr class="docutils" />
<section id="module-4-investigating-gene-flow-at-loci-of-interest-using-haplotype-networks">
<h1>Module 4 - Investigating gene flow at loci of interest using haplotype networks<a class="headerlink" href="#module-4-investigating-gene-flow-at-loci-of-interest-using-haplotype-networks" title="Permalink to this heading">#</a></h1>
<p><strong>Theme: Analysis</strong></p>
<p>In module 2 of workshop 7, we learnt about <strong>selective sweeps</strong> being shared between populations through <strong>adaptive gene flow</strong>, and in and module 3 we used hierarchical haplotype clustering to investigate and visualise haplotypes spreading between countries and species, allowing us to track gene flow at insecticide resistance loci. We also have another methodology in our armamentarium which lets us visualise the relationship between haplotypes in a different way, and that is <strong>haplotype networks</strong>. In this module we will learn how to create and interpret haplotype networks to analyse gene flow at loci associated with insecticide resistance.</p>
<p><strong>Please note: This module contains interactive plots using Plotly Dash and Cytoscape.js. These plots will not be visible on the course website, but can be generated by running the notebook for yourself. To run this notebook, click the rocket icon at the top of the page then select “Colab”.</strong></p>
<section id="learning-objectives">
<h2>Learning objectives<a class="headerlink" href="#learning-objectives" title="Permalink to this heading">#</a></h2>
<ul class="simple">
<li><p>Understand the theory behind haplotype networks.</p></li>
<li><p>Use the malariagen_data python package to build haplotype networks for genomic regions with evidence of adaptive gene flow.</p></li>
<li><p>Learn how to interpret haplotype networks to find evidence for gene flow events between countries or species.</p></li>
</ul>
</section>
<section id="lecture">
<h2>Lecture<a class="headerlink" href="#lecture" title="Permalink to this heading">#</a></h2>
<section id="english">
<h3>English<a class="headerlink" href="#english" title="Permalink to this heading">#</a></h3>
<div class="cell tag_remove-input docutils container">
<div class="cell_output docutils container">
<div class="output text_html"><iframe width="560" height="315" src="https://www.youtube.com/embed/C8LRBAq9Hpw" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
</div></div>
</div>
</section>
<section id="francais">
<h3>Français<a class="headerlink" href="#francais" title="Permalink to this heading">#</a></h3>
<div class="cell tag_remove-input docutils container">
<div class="cell_output docutils container">
<div class="output text_html"><iframe width="560" height="315" src="https://www.youtube.com/embed/PJe1wSSo19Y" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
</div></div>
</div>
<p>Please note that the code in the cells below might differ from that shown in the video. This can happen because Python packages and their dependencies change due to updates, necessitating tweaks to the code.</p>
</section>
</section>
<section id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this heading">#</a></h2>
<p>First, let’s begin by installing and importing some Python packages, and configuring access to <em>Anopheles</em> genomic data from the MalariaGEN Ag3.0 data resource.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">pip</span> install -q --no-warn-conflicts malariagen_data dash
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">malariagen_data</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">dash</span>
</pre></div>
</div>
</div>
</div>
<p>Note that authentication is required to access data through the package, please follow the instructions <a class="reference external" href="https://malariagen.github.io/vector-data/vobs/vobs-data-access.html">here</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span> <span class="o">=</span> <span class="n">malariagen_data</span><span class="o">.</span><span class="n">Ag3</span><span class="p">()</span>
<span class="n">ag3</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/javascript">'use strict';
(function(root) {
  function now() {
    return new Date();
  }

  const force = true;

  if (typeof root._bokeh_onload_callbacks === "undefined" || force === true) {
    root._bokeh_onload_callbacks = [];
    root._bokeh_is_loading = undefined;
  }

const JS_MIME_TYPE = 'application/javascript';
  const HTML_MIME_TYPE = 'text/html';
  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';
  const CLASS_NAME = 'output_bokeh rendered_html';

  /**
   * Render data to the DOM node
   */
  function render(props, node) {
    const script = document.createElement("script");
    node.appendChild(script);
  }

  /**
   * Handle when an output is cleared or removed
   */
  function handleClearOutput(event, handle) {
    function drop(id) {
      const view = Bokeh.index.get_by_id(id)
      if (view != null) {
        view.model.document.clear()
        Bokeh.index.delete(view)
      }
    }

    const cell = handle.cell;

    const id = cell.output_area._bokeh_element_id;
    const server_id = cell.output_area._bokeh_server_id;

    // Clean up Bokeh references
    if (id != null) {
      drop(id)
    }

    if (server_id !== undefined) {
      // Clean up Bokeh references
      const cmd_clean = "from bokeh.io.state import curstate; print(curstate().uuid_to_server['" + server_id + "'].get_sessions()[0].document.roots[0]._id)";
      cell.notebook.kernel.execute(cmd_clean, {
        iopub: {
          output: function(msg) {
            const id = msg.content.text.trim()
            drop(id)
          }
        }
      });
      // Destroy server and session
      const cmd_destroy = "import bokeh.io.notebook as ion; ion.destroy_server('" + server_id + "')";
      cell.notebook.kernel.execute(cmd_destroy);
    }
  }

  /**
   * Handle when a new output is added
   */
  function handleAddOutput(event, handle) {
    const output_area = handle.output_area;
    const output = handle.output;

    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only
    if ((output.output_type != "display_data") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {
      return
    }

    const toinsert = output_area.element.find("." + CLASS_NAME.split(' ')[0]);

    if (output.metadata[EXEC_MIME_TYPE]["id"] !== undefined) {
      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];
      // store reference to embed id on output_area
      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE]["id"];
    }
    if (output.metadata[EXEC_MIME_TYPE]["server_id"] !== undefined) {
      const bk_div = document.createElement("div");
      bk_div.innerHTML = output.data[HTML_MIME_TYPE];
      const script_attrs = bk_div.children[0].attributes;
      for (let i = 0; i < script_attrs.length; i++) {
        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);
        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent
      }
      // store reference to server id on output_area
      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE]["server_id"];
    }
  }

  function register_renderer(events, OutputArea) {

    function append_mime(data, metadata, element) {
      // create a DOM node to render to
      const toinsert = this.create_output_subarea(
        metadata,
        CLASS_NAME,
        EXEC_MIME_TYPE
      );
      this.keyboard_manager.register_events(toinsert);
      // Render to node
      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};
      render(props, toinsert[toinsert.length - 1]);
      element.append(toinsert);
      return toinsert
    }

    /* Handle when an output is cleared or removed */
    events.on('clear_output.CodeCell', handleClearOutput);
    events.on('delete.Cell', handleClearOutput);

    /* Handle when a new output is added */
    events.on('output_added.OutputArea', handleAddOutput);

    /**
     * Register the mime type and append_mime function with output_area
     */
    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {
      /* Is output safe? */
      safe: true,
      /* Index of renderer in `output_area.display_order` */
      index: 0
    });
  }

  // register the mime type if in Jupyter Notebook environment and previously unregistered
  if (root.Jupyter !== undefined) {
    const events = require('base/js/events');
    const OutputArea = require('notebook/js/outputarea').OutputArea;

    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {
      register_renderer(events, OutputArea);
    }
  }
  if (typeof (root._bokeh_timeout) === "undefined" || force === true) {
    root._bokeh_timeout = Date.now() + 5000;
    root._bokeh_failed_load = false;
  }

  const NB_LOAD_WARNING = {'data': {'text/html':
     "<div style='background-color: #fdd'>\n"+
     "<p>\n"+
     "BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \n"+
     "may be due to a slow or bad network connection. Possible fixes:\n"+
     "</p>\n"+
     "<ul>\n"+
     "<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\n"+
     "<li>use INLINE resources instead, as so:</li>\n"+
     "</ul>\n"+
     "<code>\n"+
     "from bokeh.resources import INLINE\n"+
     "output_notebook(resources=INLINE)\n"+
     "</code>\n"+
     "</div>"}};

  function display_loaded(error = null) {
    const el = document.getElementById(null);
    if (el != null) {
      const html = (() => {
        if (typeof root.Bokeh === "undefined") {
          if (error == null) {
            return "BokehJS is loading ...";
          } else {
            return "BokehJS failed to load.";
          }
        } else {
          const prefix = `BokehJS ${root.Bokeh.version}`;
          if (error == null) {
            return `${prefix} successfully loaded.`;
          } else {
            return `${prefix} <b>encountered errors</b> while loading and may not function as expected.`;
          }
        }
      })();
      el.innerHTML = html;

      if (error != null) {
        const wrapper = document.createElement("div");
        wrapper.style.overflow = "auto";
        wrapper.style.height = "5em";
        wrapper.style.resize = "vertical";
        const content = document.createElement("div");
        content.style.fontFamily = "monospace";
        content.style.whiteSpace = "pre-wrap";
        content.style.backgroundColor = "rgb(255, 221, 221)";
        content.textContent = error.stack ?? error.toString();
        wrapper.append(content);
        el.append(wrapper);
      }
    } else if (Date.now() < root._bokeh_timeout) {
      setTimeout(() => display_loaded(error), 100);
    }
  }

  function run_callbacks() {
    try {
      root._bokeh_onload_callbacks.forEach(function(callback) {
        if (callback != null)
          callback();
      });
    } finally {
      delete root._bokeh_onload_callbacks
    }
    console.debug("Bokeh: all callbacks have finished");
  }

  function load_libs(css_urls, js_urls, callback) {
    if (css_urls == null) css_urls = [];
    if (js_urls == null) js_urls = [];

    root._bokeh_onload_callbacks.push(callback);
    if (root._bokeh_is_loading > 0) {
      console.debug("Bokeh: BokehJS is being loaded, scheduling callback at", now());
      return null;
    }
    if (js_urls == null || js_urls.length === 0) {
      run_callbacks();
      return null;
    }
    console.debug("Bokeh: BokehJS not loaded, scheduling load and callback at", now());
    root._bokeh_is_loading = css_urls.length + js_urls.length;

    function on_load() {
      root._bokeh_is_loading--;
      if (root._bokeh_is_loading === 0) {
        console.debug("Bokeh: all BokehJS libraries/stylesheets loaded");
        run_callbacks()
      }
    }

    function on_error(url) {
      console.error("failed to load " + url);
    }

    for (let i = 0; i < css_urls.length; i++) {
      const url = css_urls[i];
      const element = document.createElement("link");
      element.onload = on_load;
      element.onerror = on_error.bind(null, url);
      element.rel = "stylesheet";
      element.type = "text/css";
      element.href = url;
      console.debug("Bokeh: injecting link tag for BokehJS stylesheet: ", url);
      document.body.appendChild(element);
    }

    for (let i = 0; i < js_urls.length; i++) {
      const url = js_urls[i];
      const element = document.createElement('script');
      element.onload = on_load;
      element.onerror = on_error.bind(null, url);
      element.async = false;
      element.src = url;
      console.debug("Bokeh: injecting script tag for BokehJS library: ", url);
      document.head.appendChild(element);
    }
  };

  function inject_raw_css(css) {
    const element = document.createElement("style");
    element.appendChild(document.createTextNode(css));
    document.body.appendChild(element);
  }

  const js_urls = ["https://cdn.bokeh.org/bokeh/release/bokeh-3.6.3.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.6.3.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.6.3.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.6.3.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.6.3.min.js"];
  const css_urls = [];

  const inline_js = [    function(Bokeh) {
      Bokeh.set_log_level("info");
    },
function(Bokeh) {
    }
  ];

  function run_inline_js() {
    if (root.Bokeh !== undefined || force === true) {
      try {
            for (let i = 0; i < inline_js.length; i++) {
      inline_js[i].call(root, root.Bokeh);
    }

      } catch (error) {throw error;
      }} else if (Date.now() < root._bokeh_timeout) {
      setTimeout(run_inline_js, 100);
    } else if (!root._bokeh_failed_load) {
      console.log("Bokeh: BokehJS failed to load within specified timeout.");
      root._bokeh_failed_load = true;
    } else if (force !== true) {
      const cell = $(document.getElementById(null)).parents('.cell').data().cell;
      cell.output_area.append_execute_result(NB_LOAD_WARNING)
    }
  }

  if (root._bokeh_is_loading === 0) {
    console.debug("Bokeh: BokehJS loaded, going straight to plotting");
    run_inline_js();
  } else {
    load_libs(css_urls, js_urls, function() {
      console.debug("Bokeh: BokehJS plotting callback run at", now());
      run_inline_js();
    });
  }
}(window));</script><div class="output text_html">
            <table class="malariagen-ag3">
                <thead>
                    <tr>
                        <th style="text-align: left" colspan="2">MalariaGEN Ag3 API client</th>
                    </tr>
                    <tr><td colspan="2" style="text-align: left">
                        Please note that data are subject to terms of use,
                        for more information see <a href="https://www.malariagen.net/data">
                        the MalariaGEN website</a> or contact support@malariagen.net.
                        See also the <a href="https://malariagen.github.io/malariagen-data-python/v15.2.0/Ag3.html">Ag3 API docs</a>.
                    </td></tr>
                </thead>
                <tbody>
                    <tr>
                        <th style="text-align: left">
                            Storage URL
                        </th>
                        <td>gs://vo_agam_release_master_us_central1</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            Data releases available
                        </th>
                        <td>3.0, 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8, 3.9, 3.10, 3.11, 3.12, 3.13, 3.14</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            Results cache
                        </th>
                        <td>None</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            Cohorts analysis
                        </th>
                        <td>20250131</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            AIM analysis
                        </th>
                        <td>20220528</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            Site filters analysis
                        </th>
                        <td>dt_20200416</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            Software version
                        </th>
                        <td>malariagen_data 15.2.0</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            Client location
                        </th>
                        <td>Iowa, United States (Google Cloud us-central1)</td>
                    </tr>
                </tbody>
            </table>
        </div></div>
</div>
</section>
<section id="haplotype-networks">
<h2>Haplotype networks<a class="headerlink" href="#haplotype-networks" title="Permalink to this heading">#</a></h2>
<p>We have an experimental function for building haplotype networks in the <code class="docutils literal notranslate"><span class="pre">malariagen_data</span></code> python package (we’re still ironing out some bugs!). Let’s have a look at the doc string for the <code class="docutils literal notranslate"><span class="pre">ag3.plot_haplotype_network()</span></code> function by using “?”.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ag3.plot_haplotype_network<span class="o">?</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Red">Signature:</span>
ag3.plot_haplotype_network(
    region: Union[str, malariagen_data.util.Region, Mapping, List[Union[str, malariagen_data.util.Region, Mapping]], Tuple[Union[str, malariagen_data.util.Region, Mapping], ...]],
    analysis: str = <span class=" -Color -Color-Yellow">&#39;default&#39;</span>,
    sample_sets: Union[str, Sequence[str], NoneType] = <span class=" -Color -Color-Bold -Color-Bold-C28">None</span>,
    sample_query: Optional[str] = <span class=" -Color -Color-Bold -Color-Bold-C28">None</span>,
    sample_query_options: Optional[dict] = <span class=" -Color -Color-Bold -Color-Bold-C28">None</span>,
    max_dist: int = <span class=" -Color -Color-Green">2</span>,
    color: Union[str, Mapping, NoneType] = <span class=" -Color -Color-Bold -Color-Bold-C28">None</span>,
    color_discrete_sequence: Optional[List] = <span class=" -Color -Color-Bold -Color-Bold-C28">None</span>,
    color_discrete_map: Optional[Mapping] = <span class=" -Color -Color-Bold -Color-Bold-C28">None</span>,
    category_orders: Union[List, Mapping, NoneType] = <span class=" -Color -Color-Bold -Color-Bold-C28">None</span>,
    node_size_factor: int = <span class=" -Color -Color-Green">50</span>,
    layout: str = <span class=" -Color -Color-Yellow">&#39;cose&#39;</span>,
    layout_params: Optional[Mapping] = <span class=" -Color -Color-Bold -Color-Bold-C28">None</span>,
    server_port: Optional[int] = <span class=" -Color -Color-Bold -Color-Bold-C28">None</span>,
    server_mode: Optional[Literal[<span class=" -Color -Color-Yellow">&#39;inline&#39;</span>, <span class=" -Color -Color-Yellow">&#39;external&#39;</span>, <span class=" -Color -Color-Yellow">&#39;jupyterlab&#39;</span>, <span class=" -Color -Color-Yellow">&#39;tab&#39;</span>]] = <span class=" -Color -Color-Yellow">&#39;inline&#39;</span>,
    height: int = <span class=" -Color -Color-Green">600</span>,
    width: Union[int, str, NoneType] = <span class=" -Color -Color-Yellow">&#39;100%&#39;</span>,
    serve_scripts_locally: bool = <span class=" -Color -Color-Bold -Color-Bold-C28">True</span>,
    chunks: Union[int, str, Tuple[Union[int, str], ...], Callable[[Tuple[int, ...]], Union[int, str, Tuple[Union[int, str], ...]]]] = <span class=" -Color -Color-Yellow">&#39;native&#39;</span>,
    inline_array: bool = <span class=" -Color -Color-Bold -Color-Bold-C28">True</span>,
)
<span class=" -Color -Color-Red">Docstring:</span>
Construct a median-joining haplotype network and display it using
Cytoscape.

A haplotype network provides a visualisation of the genetic distance
between haplotypes. Each node in the network represents a unique
haplotype. The size (area) of the node is scaled by the number of times
that unique haplotype was observed within the selected samples. A
connection between two nodes represents a single SNP difference between
the corresponding haplotypes.

Parameters
----------
region : str or Region or Mapping or list of str or Region or Mapping or tuple of str or Region or Mapping
    Region of the reference genome. Can be a contig name, region string
    (formatted like &quot;{contig}:{start}-{end}&quot;), or identifier of a genome
    feature such as a gene or transcript. Can also be a sequence (e.g.,
    list) of regions.
analysis : str, optional, default: &#39;default&#39;
    Which haplotype phasing analysis to use. See the
    `phasing_analysis_ids` property for available values.
sample_sets : sequence of str or str or None, optional
    List of sample sets and/or releases. Can also be a single sample set
    or release.
sample_query : str or None, optional
    A pandas query string to be evaluated against the sample metadata, to
    select samples to be included in the returned data.
sample_query_options : dict or None, optional
    A dictionary of arguments that will be passed through to pandas
    query() or eval(), e.g. parser, engine, local_dict, global_dict,
    resolvers.
max_dist : int, optional, default: 2
    Join network components up to a maximum distance of 2 SNP differences.
color : str or Mapping or None, optional
    Name of variable to use to color the markers.
color_discrete_sequence : List or None, optional
    Provide a list of colours to use.
color_discrete_map : Mapping or None, optional
    Provide an explicit mapping from values to colours.
category_orders : List or Mapping or None, optional
    Control the order in which values appear in the legend.
node_size_factor : int, optional, default: 50
    Control the sizing of nodes.
layout : str, optional, default: &#39;cose&#39;
    Name of the network layout to use to position nodes.
layout_params : Mapping or None, optional
    Additional parameters to the layout algorithm.
server_port : int or None, optional
    Manually override the port on which the Dash app will run.
server_mode : {&#39;inline&#39;, &#39;external&#39;, &#39;jupyterlab&#39;, &#39;tab&#39;} or None, optional, default: &#39;inline&#39;
    Controls how the Jupyter Dash app will be launched. See
    https://dash.plotly.com/dash-in-jupyter#display-modes for more
    information.
height : int, optional, default: 600
    Height of the Dash app in pixels (px).
width : int or str or None, optional, default: &#39;100%&#39;
    Width of the Dash app.
serve_scripts_locally : bool, optional, default: True
    If True, serve scripts locally, otherwise use CDN.
chunks : int or str or tuple of int or str or Callable[[typing.Tuple[int, ...]], int or str or tuple of int or str], optional, default: &#39;native&#39;
    Define how input data being read from zarr should be divided into
    chunks for a dask computation. If &#39;native&#39;, use underlying zarr
    chunks. If a string specifying a target memory size, e.g., &#39;300 MiB&#39;,
    resize chunks in arrays with more than one dimension to match this
    size. If &#39;auto&#39;, let dask decide chunk size.  If &#39;ndauto&#39;, let dask
    decide chunk size but only for arrays with more than one dimension. If
    &#39;ndauto0&#39;, as &#39;ndauto&#39; but only vary the first chunk dimension. If
    &#39;ndauto1&#39;, as &#39;ndauto&#39; but only vary the second chunk dimension. If
    &#39;ndauto01&#39;, as &#39;ndauto&#39; but only vary the first and second chunk
    dimensions. Also, can be a tuple of integers, or a callable which
    accepts the native chunks as a single argument and returns a valid
    dask chunks value.
inline_array : bool, optional, default: True
    Passed through to dask `from_array()`.
<span class=" -Color -Color-Red">File:</span>      /home/conda/developer/35674e27b19f7c625ba32a1b88449ff45c90b40edb90a065b66c5a9a5388f41c-20250421-195247-360829-93-training-nb-maintenance-mgen-15.2.0/lib/python3.11/site-packages/malariagen_data/anopheles.py
<span class=" -Color -Color-Red">Type:</span>      method
</pre></div>
</div>
</div>
</div>
<p>The required parameters, <code class="docutils literal notranslate"><span class="pre">region</span></code> and <code class="docutils literal notranslate"><span class="pre">analysis</span></code> are familiar to us by now, but there are a some new optional parameters here that might be useful to understand.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">max_dist=</span></code> This is the maximum distance in <em>n</em> SNPs where the algorithm will join haplotype nodes together as one network, after which it will split the nodes in to separate networks. The default here is 2 SNPs. We will see an example of how this works shortly.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">color=</span></code> Identifies a column in the sample metadata which determines the colour of pie chart segments within nodes. <em>e.g.</em> we could use “country” or “taxon” columns.</p></li>
</ul>
<p>It is also worth knowing that if you get <strong>403 errors</strong> instead of figures, try reloading your browser tab or restart your browser.</p>
<p>Let’s build a haplotype network around <em>Kdr</em> locus in the <em>Vgsc</em> gene for all the <em>An. coluzzii</em> samples in Ag3.0, and look at the output.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span><span class="o">.</span><span class="n">plot_haplotype_network</span><span class="p">(</span>
    <span class="n">region</span><span class="o">=</span><span class="s1">&#39;2L:2,400,000-2,440,000&#39;</span><span class="p">,</span>
    <span class="n">analysis</span><span class="o">=</span><span class="s1">&#39;gamb_colu&#39;</span><span class="p">,</span>
    <span class="n">sample_sets</span><span class="o">=</span><span class="s1">&#39;3.0&#39;</span><span class="p">,</span>
    <span class="n">sample_query</span><span class="o">=</span> <span class="s1">&#39;taxon == &quot;coluzzii&quot;&#39;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;country&quot;</span><span class="p">,</span>
    <span class="n">max_dist</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>N.B., you will need to run this notebook for yourself in colab in order for the haplotype network to work.</p>
<section id="technical-interpretation">
<h3>Technical interpretation<a class="headerlink" href="#technical-interpretation" title="Permalink to this heading">#</a></h3>
<p>The output figure represents identical haplotypes as <strong>nodes</strong> (circles), the larger the node, the greater the number of identical haplotypes. In this example, the <code class="docutils literal notranslate"><span class="pre">color</span></code> parameter has been set to the “country” column of the haplotype metadata, so the node colour proportions represent the countries where that haplotype was present in our data set. We often also set the <code class="docutils literal notranslate"><span class="pre">color</span></code> parameter to “taxon”, in which case, node colour proportion represents the species which carried that haplotype. If we click or tap a node, information about its composition is provided at the bottom of the figure. This information relates the the <code class="docutils literal notranslate"><span class="pre">color</span></code> parameter, <em>i.e.</em> if <code class="docutils literal notranslate"><span class="pre">color=taxon</span></code>, the information will be about species composition. Nodes can also be moved to aid visualisation (<code class="docutils literal notranslate"><span class="pre">shift</span> <span class="pre">+</span> <span class="pre">click</span></code> to select multiple), and the whole figure can be zoomed in and out using the mouse wheel.</p>
<p>The <strong>edges</strong> (lines) that join nodes together represent the genetic distance between the nodes in SNPs. Each edge is a single SNP difference.</p>
<p>In the <em>Vgsc</em> example, with the <code class="docutils literal notranslate"><span class="pre">max_dist</span></code> (maximum distance) parameter set to two, we se each node in these networks is separated by a maximum genetic distance of 2 SNPs, shown as two edges with a small grey node in the middle. If we click on a small grey node we see “No. haplotypes: 0”. The evolutionary relationship between these haplotype nodes either side of the small node has been inferred as we don’t actually have the intermediate haplotype (with only one SNP difference) in our data set. Let’s increase the <code class="docutils literal notranslate"><span class="pre">max_dist</span></code> parameter and see what happens.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span><span class="o">.</span><span class="n">plot_haplotype_network</span><span class="p">(</span>
    <span class="n">region</span><span class="o">=</span><span class="s1">&#39;2L:2,400,000-2,440,000&#39;</span><span class="p">,</span>
    <span class="n">analysis</span><span class="o">=</span><span class="s1">&#39;gamb_colu&#39;</span><span class="p">,</span>
    <span class="n">sample_sets</span><span class="o">=</span><span class="s1">&#39;3.0&#39;</span><span class="p">,</span>
    <span class="n">sample_query</span><span class="o">=</span> <span class="s1">&#39;taxon == &quot;coluzzii&quot;&#39;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;country&quot;</span><span class="p">,</span>
    <span class="n">max_dist</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>N.B., you will need to run this notebook for yourself in colab in order for the haplotype network to work.</p>
<p>If we increase the <code class="docutils literal notranslate"><span class="pre">max_dist</span></code> parameter to four SNPs, we see haplotypes linked together by up to three “intermediate” node, the network has found connections between more distantly related haplotypes. We generally use <code class="docutils literal notranslate"><span class="pre">max_dist=2</span></code>, mainly because we’re interested here only in groups of very closely related haplotypes, which are therefore very likely to share a recent common origin.</p>
<p>To aid interpretation, <strong>only networks containing more than one haplotype are represented in the output figure</strong>. This means that if a particular haplotype is found only once in our data set (<em>i.e.</em> a singleton haplotype), and that haplotype is more than the <code class="docutils literal notranslate"><span class="pre">max_dist</span></code> number of SNPs different from any other haplotype, it not be shown.</p>
</section>
<section id="evolutionary-interpretation">
<h3>Evolutionary interpretation<a class="headerlink" href="#evolutionary-interpretation" title="Permalink to this heading">#</a></h3>
<p>Let’s return to the <em>Vgsc</em> network example with <code class="docutils literal notranslate"><span class="pre">max_dist=2</span></code> and try to interpret the evolutionary inference with respect to insecticide resistance.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span><span class="o">.</span><span class="n">plot_haplotype_network</span><span class="p">(</span>
    <span class="n">region</span><span class="o">=</span><span class="s1">&#39;2L:2,400,000-2,440,000&#39;</span><span class="p">,</span>
    <span class="n">analysis</span><span class="o">=</span><span class="s1">&#39;gamb_colu&#39;</span><span class="p">,</span>
    <span class="n">sample_sets</span><span class="o">=</span><span class="s1">&#39;3.0&#39;</span><span class="p">,</span>
    <span class="n">sample_query</span><span class="o">=</span> <span class="s1">&#39;taxon == &quot;coluzzii&quot;&#39;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;country&quot;</span><span class="p">,</span>
    <span class="n">max_dist</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>N.B., you will need to run this notebook for yourself in colab in order for the haplotype network to work.</p>
<ul class="simple">
<li><p>When a fitness increasing mutation occurs, haplotype it resides on comes under selection and it increases in frequency. So, here we’re interested especially in the large nodes, because these are very likely to be <strong>high frequency haplotypes experiencing positive selection</strong>.</p></li>
<li><p>During a selective sweep, some additional mutations will occur on the original haplotype. These could be neutral or they could be secondary functional mutations (<em>i.e.</em> further increasing fitness), but key point is that these additional mutations occur, and give us a network of closely related haplotypes. We can infer that each network of connected haplotypes have a recent common origin.</p></li>
<li><p>If a network of connected haplotypes contains haplotypes from two or more countries, this is evidence for adaptive gene flow. Particularly a large node containing haplotypes from multiple countries, this is very strong evidence for adaptive gene flow because exactly the same haplotype is found across countries.</p></li>
<li><p>Likewise, if a network of connected haplotypes contains haplotypes from two or more species, this is also evidence for a type of adaptive gene flow called <strong>adaptive introgression</strong>. Again, a large node containing haplotypes from multiple species, is very strong evidence for adaptive gene flow.</p></li>
</ul>
<p>The most striking feature of this <em>Vgsc</em> network is the very large and complex network at the top the figure. The larger node colours show that there are several high frequency haplotypes, closely related to each other, shared across countries. This is a classic signal of adaptive gene flow and in this case, we know from previous research, the haplotypes all carry the L995F <em>kdr</em> mutations that confers resistance to pyrethroids and DDT. Furthermore, the multiple large nodes suggest that secondary mutations have recently occurred on an already selected haplotype, then these new haplotypes have increased in frequency, likely also under selection.</p>
<p>We can also see some smaller complex networks, which suggest potential selective sweeps, two of which show evidence of adaptive gene flow between countries (sharing of identical haplotypes).</p>
<p>The smallest networks are fairly uninformative, and could be considered noise.</p>
<p>Now we have an idea how to run and interpret haplotype networks, let’s look at some insecticide resistance loci we looked at in module 3.</p>
</section>
</section>
<section id="exercise-1">
<h2>Exercise 1<a class="headerlink" href="#exercise-1" title="Permalink to this heading">#</a></h2>
<section id="id1">
<h3>English<a class="headerlink" href="#id1" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Repeat the <em>Vgsc</em> network analysis, but this time investigate adaptive gene flow (introgression) between <em>An. gambiae</em> and <em>coluzzii</em>. <em>HINT</em>: <code class="docutils literal notranslate"><span class="pre">color=?????</span></code>.</p></li>
<li><p>Is there evidence of adaptive gene flow between species?</p></li>
</ul>
</section>
<section id="id2">
<h3>Français<a class="headerlink" href="#id2" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Répéter l’analyse des réseaux pour <em>Vgsc</em> mais cette fois étudier le flux génétique adaptatif (introgression) entre <em>An. gambiae</em> et <em>coluzzii</em>. <em>INDICE</em>: <code class="docutils literal notranslate"><span class="pre">color=?????</span></code>.</p></li>
<li><p>Y a-t-il des indications de flux génétique adaptatif entre espèces?</p></li>
</ul>
</section>
</section>
<section id="example-ace1">
<h2>Example: <em>Ace1</em><a class="headerlink" href="#example-ace1" title="Permalink to this heading">#</a></h2>
<p>We’ve heard about the <em>Ace1</em> genes involvement in organophosphate insecticide resistance, and in the last module, we’ve also seen the results of visualising the relationships between haplotypes at this locus using <strong>hierarchical clustering</strong>. Let’s first remind ourselved what that looked like for <em>Ace1</em> with <em>An. gambiae</em> haplotypes from from Ghana and Burkina Faso..</p>
<section id="ace1-hierarchical-clustering-re-cap">
<h3><em>Ace1</em> hierarchical clustering re-cap<a class="headerlink" href="#ace1-hierarchical-clustering-re-cap" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span><span class="o">.</span><span class="n">plot_haplotype_clustering</span><span class="p">(</span>
    <span class="n">region</span><span class="o">=</span><span class="s1">&#39;2R:3,470,000-3,490,000&#39;</span><span class="p">,</span>
    <span class="n">analysis</span><span class="o">=</span><span class="s1">&#39;gamb_colu&#39;</span><span class="p">,</span>
    <span class="n">sample_sets</span><span class="o">=</span><span class="s1">&#39;3.0&#39;</span><span class="p">,</span>
    <span class="n">sample_query</span><span class="o">=</span> <span class="s1">&#39;country in [&quot;Ghana&quot;, &quot;Burkina Faso&quot;] &amp; taxon == &quot;gambiae&quot;&#39;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s1">&#39;country&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>N.B., you will need to run this notebook for yourself in colab in order for the haplotype network to work.</p>
<p>Towards the right hand side of the dendrogram, we can see a region where there are haplotypes closely related to each other. By zooming in we can find that this distance is &lt;5 SNPs between haplotypes, rather than &gt;50 SNPs distance we see elsewhere in the figure. This pattern is suggestive of a selective sweep, and as it includes multiple countries, is also suggestive of <strong>adaptive gene flow</strong>.</p>
<p>Adaptive gene flow is where a fitness improving mutation increases in frequency and spreads across space, between populations and/or between different species. Here we are detecting the increase and spread of the haplotype which carries this locus under positive selection. Given the importance of this gene to insecticide action and the high evolutionary pressure that these chemicals put on mosquitoes, this signal may well be driven by a resistance phenotype associated locus.</p>
<p>Let’s investigate the same region but using haplotype networks.</p>
</section>
<section id="ace1-haplotype-networks">
<h3><em>Ace1</em> haplotype networks<a class="headerlink" href="#ace1-haplotype-networks" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span><span class="o">.</span><span class="n">plot_haplotype_network</span><span class="p">(</span>
    <span class="n">region</span><span class="o">=</span><span class="s1">&#39;2R:3,470,000-3,490,000&#39;</span><span class="p">,</span>
    <span class="n">analysis</span><span class="o">=</span><span class="s1">&#39;gamb_colu&#39;</span><span class="p">,</span>
    <span class="n">sample_sets</span><span class="o">=</span><span class="s1">&#39;3.0&#39;</span><span class="p">,</span>
    <span class="n">sample_query</span><span class="o">=</span> <span class="s1">&#39;country in [&quot;Ghana&quot;, &quot;Burkina Faso&quot;] &amp; taxon == &quot;gambiae&quot;&#39;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;country&quot;</span><span class="p">,</span>
    <span class="n">max_dist</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>N.B., you will need to run this notebook for yourself in colab in order for the haplotype network to work.</p>
<p>Looking back at the hierarchical clustering analysis of <em>Ace1</em>, the vast majority of the dendrogram leaves (haplotypes) have long branches (genetic distances) to their nearest neighbours. All of those singleton haplotypes, more that 2 SNPs distance from any other haplotypes, are filtered out in the network analysis as we are only interested in high frequency haplotypes, and networks of closely related haplotypes, because they are likely to be haplotypes under recent positive selection. What remains are the haplotypes which make us the networks we can see above, a much simplified picture of the genetic relationships.</p>
<p>The largest network corresponds to the selective sweep signal we saw in the clustering analysis. Clicking on the largest node, we see it contains 15 identical haplotypes, 11 from Burkina Faso and four from Ghana. As expected, this corresponds perfectly with what we see if we zoom in to the far right side of the clustering figure and roll over the leaves with our mouse, further evidence this haplotype has spread, under selection between countries. Several larger nodes, closely related to each other gives us evidence that the selection is recent (as we might expect if driven by malaria control insecticidal pressures).</p>
</section>
</section>
<section id="exercise-2">
<h2>Exercise 2<a class="headerlink" href="#exercise-2" title="Permalink to this heading">#</a></h2>
<section id="id3">
<h3>English<a class="headerlink" href="#id3" title="Permalink to this heading">#</a></h3>
<p>For <em>Ace1</em> add <em>coluzzii</em> samples as well, colour the nodes by taxon and take another look at these networks. <em>HINT</em> <code class="docutils literal notranslate"><span class="pre">sample_query=</span> <span class="pre">'country</span> <span class="pre">in</span> <span class="pre">[&quot;Ghana&quot;,</span> <span class="pre">&quot;Burkina</span> <span class="pre">Faso&quot;]</span> <span class="pre">&amp;</span> <span class="pre">taxon</span> <span class="pre">in</span> <span class="pre">[&quot;gambiae&quot;,</span> <span class="pre">&quot;coluzzii&quot;]'</span></code></p>
<ul class="simple">
<li><p>is there evidence of adaptive introgression of haplotypes between species at the <em>Ace1</em> locus?</p></li>
</ul>
</section>
<section id="id4">
<h3>Français<a class="headerlink" href="#id4" title="Permalink to this heading">#</a></h3>
<p>Pour <em>Ace1</em> ajouter les <em>coluzzii</em>, colorer les noeuds par taxon et regarder à nouveau les réseaux. <em>INDICE</em> <code class="docutils literal notranslate"><span class="pre">sample_query=</span> <span class="pre">'country</span> <span class="pre">in</span> <span class="pre">[&quot;Ghana&quot;,</span> <span class="pre">&quot;Burkina</span> <span class="pre">Faso&quot;]</span> <span class="pre">&amp;</span> <span class="pre">taxon</span> <span class="pre">in</span> <span class="pre">[&quot;gambiae&quot;,</span> <span class="pre">&quot;coluzzii&quot;]'</span></code></p>
<ul class="simple">
<li><p>Y a-t-il des indications d’introgression adaptive entre espèces au locus <em>Ace1</em> ?</p></li>
</ul>
</section>
</section>
<section id="example-gste2">
<h2>Example: <em>Gste2</em><a class="headerlink" href="#example-gste2" title="Permalink to this heading">#</a></h2>
<p>We have seen that Gste2 gene is associated with insecticide resistance, and we saw evidence for adaptive gene flow at this locus using hierarchical clustering. Lets now perform a haplotype networking analysis at this gene and compare the results.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span><span class="o">.</span><span class="n">plot_haplotype_clustering</span><span class="p">(</span>
    <span class="n">region</span><span class="o">=</span><span class="s1">&#39;3R:28,590,000-28,600,000&#39;</span><span class="p">,</span>
    <span class="n">analysis</span><span class="o">=</span><span class="s1">&#39;gamb_colu&#39;</span><span class="p">,</span>
    <span class="n">sample_sets</span><span class="o">=</span><span class="s1">&#39;3.0&#39;</span><span class="p">,</span>
    <span class="n">sample_query</span><span class="o">=</span> <span class="s1">&#39;country in [&quot;Ghana&quot;, &quot;Burkina Faso&quot;, &quot;Cote d</span><span class="se">\&#39;</span><span class="s1">Ivoire&quot;]&#39;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s1">&#39;taxon&#39;</span><span class="p">,</span>
    <span class="n">symbol</span><span class="o">=</span><span class="s1">&#39;country&#39;</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span><span class="o">.</span><span class="n">plot_haplotype_network</span><span class="p">(</span>
    <span class="n">region</span><span class="o">=</span><span class="s1">&#39;3R:28,590,000-28,600,000&#39;</span><span class="p">,</span>
    <span class="n">analysis</span><span class="o">=</span><span class="s1">&#39;gamb_colu&#39;</span><span class="p">,</span>
    <span class="n">sample_sets</span><span class="o">=</span><span class="s1">&#39;3.0&#39;</span><span class="p">,</span>
    <span class="n">sample_query</span><span class="o">=</span> <span class="s1">&#39;country in [&quot;Ghana&quot;, &quot;Burkina Faso&quot;, &quot;Cote d</span><span class="se">\&#39;</span><span class="s1">Ivoire&quot;]&#39;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;country&quot;</span><span class="p">,</span>
    <span class="n">max_dist</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span><span class="o">.</span><span class="n">plot_haplotype_network</span><span class="p">(</span>
    <span class="n">region</span><span class="o">=</span><span class="s1">&#39;3R:28,590,000-28,600,000&#39;</span><span class="p">,</span>
    <span class="n">analysis</span><span class="o">=</span><span class="s1">&#39;gamb_colu&#39;</span><span class="p">,</span>
    <span class="n">sample_sets</span><span class="o">=</span><span class="s1">&#39;3.0&#39;</span><span class="p">,</span>
    <span class="n">sample_query</span><span class="o">=</span> <span class="s1">&#39;country in [&quot;Ghana&quot;, &quot;Burkina Faso&quot;, &quot;Cote d</span><span class="se">\&#39;</span><span class="s1">Ivoire&quot;]&#39;</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s2">&quot;taxon&quot;</span><span class="p">,</span>
    <span class="n">max_dist</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>N.B., you will need to run this notebook for yourself in colab in order for the haplotype networks to work.</p>
<p>The haplotype network figures make it easy to spot the selective sweeps and gene flow between Burkina Faso and Ghana (but not Cote d’Ivoire) and the adaptive introgression of haplotypes between <em>gambiae</em> and <em>coluzzii</em> at this locus.</p>
</section>
<section id="exercise-3">
<h2>Exercise 3<a class="headerlink" href="#exercise-3" title="Permalink to this heading">#</a></h2>
<section id="id5">
<h3>English<a class="headerlink" href="#id5" title="Permalink to this heading">#</a></h3>
<p>The cytochrome P450 gene <em>Cyp9k1</em> is associated with metabolic pyrethroid resistance. Run a haplotype clustering analysis on samples from Ghana and Cote d’Ivoire on this region. <em>HINT</em>:  region=’X:15,240,000-15,250,000’</p>
<ul class="simple">
<li><p>Is there evidence of selection?</p></li>
<li><p>Is there evidence of adaptive gene flow?</p></li>
<li><p>If so, between countries and/or species?</p></li>
</ul>
</section>
<section id="id6">
<h3>Français<a class="headerlink" href="#id6" title="Permalink to this heading">#</a></h3>
<p>Le cytochrome P450 gene <em>Cyp9k1</em> est associé à la résistance métabolique aux pyrethrinoïdes. Exécuter une analyse par regroupement d’haplotypes pour les échantillons du Ghana et de la Côte d’Ivoire dans cette région. <em>INDICE</em>: region=’X:15,240,000-15,250,000’</p>
<ul class="simple">
<li><p>Y a-t-il des indications de sélection?</p></li>
<li><p>Y a-t-il des indications de flux génétique adaptatif?</p></li>
<li><p>Si oui, entre pays et/ou espèces?</p></li>
</ul>
</section>
</section>
<section id="well-done">
<h2>Well done!<a class="headerlink" href="#well-done" title="Permalink to this heading">#</a></h2>
<p>In this module we have:</p>
<ul class="simple">
<li><p>Learnt about the theory behind haplotype networks.</p></li>
<li><p>Used the malariagen_data python package to build haplotype networks for genomic regions with evidence of adaptive gene flow.</p></li>
<li><p>Learnt how to interpret haplotype networks to discern movement of insecticide resistance loci between countries and species.</p></li>
</ul>
</section>
<section id="exercises">
<h2>Exercises<a class="headerlink" href="#exercises" title="Permalink to this heading">#</a></h2>
<section id="id7">
<h3>English<a class="headerlink" href="#id7" title="Permalink to this heading">#</a></h3>
<p>Open this notebook in Google Colab and run it for yourself from top to bottom. As you run through the notebook, cell by cell, think about what each cell is doing, and try the practical exercises along the way.</p>
<p>Have go at the practical exercises, but please don’t worry if you don’t have time to do them all during the practical session, and please ask the teaching assistants for help if you are stuck.</p>
<p>Hint: To open the notebook in Google Colab, click the rocket icon at the top of the page, then select “Colab” from the drop-down menu.</p>
</section>
<section id="id8">
<h3>Français<a class="headerlink" href="#id8" title="Permalink to this heading">#</a></h3>
<p>Ouvrir ce notebook dans Google Colab et l’exécuter vous-même du début à la fin. Pendant que vous exécutez le notebook, cellule par cellule, pensez à ce que chaque cellule fait et essayez de faire les exercices quand vous les rencontrez.</p>
<p>Essayez de faire les exercices mais ne vous inquiétez pas si vous n’avez pas le temps de tout faire pendant la séance appliquée et n’hésitez pas à demander aux enseignants assistants si vous avez besoin d’aide parce que vous êtes bloqués.</p>
<p>Indice: Pour ouvrir le notebook dans Google Colab, cliquer sur l’icône de fusée au sommet de cette page puis choisissez “Colab” dans le menu déroulant.</p>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "conda-env-developer-developer-training-nb-maintenance-mgen-15.2.0-py"
        },
        kernelOptions: {
            name: "conda-env-developer-developer-training-nb-maintenance-mgen-15.2.0-py",
            path: "./workshop-7"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'conda-env-developer-developer-training-nb-maintenance-mgen-15.2.0-py'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="module-3-haplotype-clustering.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Module 3 - Investigating gene flow at loci of interest using haplotype clustering</p>
      </div>
    </a>
    <a class="right-next"
       href="../workshop-8/about.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Workshop 8 - Analysis planning and reporting</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-objectives">Learning objectives</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#lecture">Lecture</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#english">English</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#francais">Français</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup">Setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#haplotype-networks">Haplotype networks</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#technical-interpretation">Technical interpretation</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#evolutionary-interpretation">Evolutionary interpretation</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-1">Exercise 1</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">English</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Français</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-ace1">Example: <em>Ace1</em></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ace1-hierarchical-clustering-re-cap"><em>Ace1</em> hierarchical clustering re-cap</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ace1-haplotype-networks"><em>Ace1</em> haplotype networks</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-2">Exercise 2</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">English</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Français</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example-gste2">Example: <em>Gste2</em></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exercise-3">Exercise 3</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">English</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id6">Français</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#well-done">Well done!</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exercises">Exercises</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id7">English</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id8">Français</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By MalariaGEN and PAMCA
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a> This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>