

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Module 1 - Introduction to pandas DataFrames &#8212; Training course in data analysis for genomic surveillance of African malaria vectors</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/myfile.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="//cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-MML-AM_CHTML"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script type="application/vnd.jupyter.widget-state+json">{"state": {"0122747ca8cc4ef7ba7c249c3bdfd7e3": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "01feec3ce0a44d8d8dcefbd38adec76d": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "03bbbfb2510b463c9d0484c9309fdbb4": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "04686bb523004728b12862be59d0ca43": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "04b8290f1b93499aa28d012eccdf2381": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_03bbbfb2510b463c9d0484c9309fdbb4", "max": 9675, "style": "IPY_MODEL_efea2c9de56044148caab1cd7fc8018d", "value": 9675}}, "09e775321e064dc186a94213be9cd428": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "0a9bc466f51b4164b257102956ff9080": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_69fb2be1f02644729f7ad515bd05552c", "IPY_MODEL_6965c435d78f4951963d7f6679f78eae", "IPY_MODEL_4b7f8f2d17a340ca93873307754a30cc"], "layout": "IPY_MODEL_1bebd8d1accd4627b3ec2790e3e9b27c"}}, "0b759c0b529f493aaf9bd99c1984c7d0": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "1056443ce9304f6dbf1c20e449fb7794": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_b78ddc81e8c14c70b6c59cd216703779", "max": 32, "style": "IPY_MODEL_61e1a22db7394506b0fca90e9538605b", "value": 32}}, "114446d1668d40b0a5f695419aec985d": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_a0f38c1810ab4229a9bd25d23783514c", "style": "IPY_MODEL_358c11e23fdf43f9bd3e15f96e65b737", "value": " 6/6 [00:00&lt;00:00, 49.56it/s]"}}, "11e086ad1806434db236dc242c587bb4": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "121551ea3b314f3281bb296981ddaa3a": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "13d12f42b8e64833b127617fa267f816": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_d6dbfdbd0b09460799b4456ec57e7127", "max": 32, "style": "IPY_MODEL_8cd22ebdfff5403baab477a4eb37bce6", "value": 32}}, "16f86a032d2544f1b062c82b3e4fdfcc": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "17d5610bc5744ac78a017248e9024b2a": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_93d28dc09e80429b949589070a80db84", "style": "IPY_MODEL_39c12e5076284ec8b1478f4943cdbcc2", "value": "Load SNP genotypes: 100%"}}, "198865cffd8a4e23a532c729e1704615": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "1abbaaba9cf044c6b95d3ce434b21ef1": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "1bebd8d1accd4627b3ec2790e3e9b27c": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "1d82e912a75243a7a95085e4d1b013bd": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "1d85e3478d3f43d9bf1fbff545c08687": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_3606643905cf427c8d90cabea8caf0e8", "style": "IPY_MODEL_24836607042740c494da16ba6c8b229c", "value": "Compute SNP effects: 100%"}}, "1ea2e74af4f74c518d1ed8d0740b4c2b": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "2108a6c2f0e8473f9465050d08f9278c": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "22da57a2a1754af39446090815ac5218": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "24836607042740c494da16ba6c8b229c": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "24a27df073604a0d82272c87326ed191": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "270959f2f69f4db1bcdf2e885c5e8995": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_a2b8d6ef5a064619ab9f0b9e7ecb23d8", "style": "IPY_MODEL_b5e3dd8ae5544641b7e47825eb3ab086", "value": " 9675/9675 [00:00&lt;00:00, 24009.78it/s]"}}, "2718927363974000af1ed9d7ddcdf973": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "2822f1eb4f5b490aad73a7564cbd2e5a": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "2ae68bf5b37343c185d84add35e8818f": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "2eddc221216f4e85af7e5f00f40771a0": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "2f0f251f5462473dae4e2c4151e96e33": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "31a55dab84e34de7ae58647ecdf0d53c": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "3354d7c9c5f34d11bb57406d045931d1": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_16f86a032d2544f1b062c82b3e4fdfcc", "style": "IPY_MODEL_f46eed124c8740e8b5dfea671fa05a44", "value": "Compute SNP effects: 100%"}}, "355560fa88c2458f90079129ea57df2b": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_7cf8cffeaba54fa1955afeed06441fbd", "style": "IPY_MODEL_2822f1eb4f5b490aad73a7564cbd2e5a", "value": " 6/6 [00:00&lt;00:00, 46.19it/s]"}}, "358c11e23fdf43f9bd3e15f96e65b737": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "35a44715966543a580aca8be0b1a43d9": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "3606643905cf427c8d90cabea8caf0e8": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "36e2163f360e4aac9d4b526ea55d1421": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_962547f642b14d63bb2f01d4750ad3bf", "IPY_MODEL_f05d35c7b9284d8397b99c430c0284b2", "IPY_MODEL_9eacb59f381d4a0f89505f67a93475c6"], "layout": "IPY_MODEL_59c85bc9c8124c9d81c2b36e40e26a00"}}, "39c12e5076284ec8b1478f4943cdbcc2": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "3b93cb367a0042019707aa2eb2beddd9": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_1d85e3478d3f43d9bf1fbff545c08687", "IPY_MODEL_04b8290f1b93499aa28d012eccdf2381", "IPY_MODEL_3ed2ed85f8884a95ab244ebd09b2c0fd"], "layout": "IPY_MODEL_1ea2e74af4f74c518d1ed8d0740b4c2b"}}, "3d6905478253445ca29ca73f162876a6": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_7a02e1056cd9495f929dfad15737992c", "IPY_MODEL_a037168605524e8e957941a1cf28650e", "IPY_MODEL_b7c48493b45445a1bcd885bfd59004e2"], "layout": "IPY_MODEL_0b759c0b529f493aaf9bd99c1984c7d0"}}, "3ed2ed85f8884a95ab244ebd09b2c0fd": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_5c1e1a574ade4ad6af0118de5e65f161", "style": "IPY_MODEL_bcda8a6537834de1befff12992a89159", "value": " 9675/9675 [00:02&lt;00:00,  1.91s/it]"}}, "4503cc6920354eeca0f6a0b1257a1b05": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "466ca83f6c324f30ae3ca3275c3ab78b": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "4836d38622be48aab4e461f4e0c35a53": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "4b7f8f2d17a340ca93873307754a30cc": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_ba69408b986a4483b215c75086a3ea22", "style": "IPY_MODEL_f4d97bd502974cd7b455befa930858ea", "value": " 32/32 [00:01&lt;00:00, 31.70it/s]"}}, "4d178fc793fd4c67a71e1e90012e017b": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_e72011bffa6347f2a9cb820cfd611adf", "max": 6, "style": "IPY_MODEL_cf7e0dab672243f7afe4c3c47c2ddd9d", "value": 6}}, "4f75f60e4a0f43ca97fa70a239aed625": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_f1574ba01610458a8876b76f8f59f807", "max": 9675, "style": "IPY_MODEL_121551ea3b314f3281bb296981ddaa3a", "value": 9675}}, "4fa230c19e0a4a3583f0b3b1e85531ca": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "50de8fd4b8c34326a58de2c550a66048": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_ba66918cf3bd42fb8a7e481b3ed5e5e0", "style": "IPY_MODEL_2f0f251f5462473dae4e2c4151e96e33", "value": "Compute SNP effects: 100%"}}, "536be44d51e04783b7d21c550ff7c8ea": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_630da3eb815f421595ca821679ddfe78", "IPY_MODEL_1056443ce9304f6dbf1c20e449fb7794", "IPY_MODEL_7af96a4a3e3344aa86861ca09e3bd885"], "layout": "IPY_MODEL_4fa230c19e0a4a3583f0b3b1e85531ca"}}, "543c034668f342a49b6473f393943af2": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_7dff54605d454264a6466dab4bfdc9f3", "IPY_MODEL_13d12f42b8e64833b127617fa267f816", "IPY_MODEL_9b21e6cc977c414a99f9faa3421668ba"], "layout": "IPY_MODEL_f9dcd46db93b457784b840fa0def44a2"}}, "570dc7cd94bd4c308b5b80ffef8276f1": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "597023dfd6b64751bf548422d50f4ea9": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "59c85bc9c8124c9d81c2b36e40e26a00": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "5c1e1a574ade4ad6af0118de5e65f161": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "615441c5e6704292b65eaba257b563f1": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_0122747ca8cc4ef7ba7c249c3bdfd7e3", "style": "IPY_MODEL_e2b9df016be346a8a0ec7a064f7874a4", "value": "Compute allele frequencies: 100%"}}, "61e1a22db7394506b0fca90e9538605b": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "61e93db6b2494fd6a2485b99d6930ce8": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "62f26b4b68dd461698800a2286704b56": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_7d34ff7e04a340f4ae53a37b0e001246", "IPY_MODEL_6d3da9769cad496d9895901699a240ef", "IPY_MODEL_114446d1668d40b0a5f695419aec985d"], "layout": "IPY_MODEL_11e086ad1806434db236dc242c587bb4"}}, "630da3eb815f421595ca821679ddfe78": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_ea1469ca85594801a87dab85308ca6ae", "style": "IPY_MODEL_cf718db6796740f0b94f1063346fdce1", "value": "Load SNP genotypes: 100%"}}, "65aa547efcf1478dbf2840c6865af214": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "67d78a35a5724ca086ef51d8856cb74d": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "695d241857564936a6e07dcd9007860c": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "6965c435d78f4951963d7f6679f78eae": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_70cd9df80146499daf6ca33cf7007874", "max": 32, "style": "IPY_MODEL_6fb5a8db91f84d95a54ed679459626e3", "value": 32}}, "69fb2be1f02644729f7ad515bd05552c": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_2eddc221216f4e85af7e5f00f40771a0", "style": "IPY_MODEL_714aa0a45e1b4000b366d196748855dd", "value": "Load SNP genotypes: 100%"}}, "6cb7d773758d42fc9c1c6bc131b154b8": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_77f6092064a148c8a3c43d99f5de529a", "max": 32, "style": "IPY_MODEL_b4343a439bd840f493f5db9c291bf327", "value": 32}}, "6d3da9769cad496d9895901699a240ef": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_04686bb523004728b12862be59d0ca43", "max": 6, "style": "IPY_MODEL_e8e8c21b4a96459a9801835c02cea33b", "value": 6}}, "6fb5a8db91f84d95a54ed679459626e3": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "70cd9df80146499daf6ca33cf7007874": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "714aa0a45e1b4000b366d196748855dd": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "721d20b3c62343b793c9f60fdec870e2": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "75cdfbace08644edbc168eacb2531bdf": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_3354d7c9c5f34d11bb57406d045931d1", "IPY_MODEL_a42d8fc38b0e45c397757cfaacc129bb", "IPY_MODEL_b574e1bc044043ac9ca380b8863df5b7"], "layout": "IPY_MODEL_eb5c25ba5b5f42068ef12526051ba499"}}, "77f6092064a148c8a3c43d99f5de529a": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "7a02e1056cd9495f929dfad15737992c": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_f9e452b4cc654a499daa5b729ad45806", "style": "IPY_MODEL_61e93db6b2494fd6a2485b99d6930ce8", "value": "Load SNP genotypes: 100%"}}, "7af96a4a3e3344aa86861ca09e3bd885": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_f765d760859d410eacc2c7c0c567aac1", "style": "IPY_MODEL_bceb261c28dc46c6aac5c68b3836b22e", "value": " 32/32 [00:00&lt;00:00, 81.37it/s]"}}, "7cf8cffeaba54fa1955afeed06441fbd": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "7d34ff7e04a340f4ae53a37b0e001246": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_d165dc10dd004f7088c2d8b9aa0625d7", "style": "IPY_MODEL_e4603bc7527c42e58eacc29ea1930df8", "value": "Compute allele frequencies: 100%"}}, "7dff54605d454264a6466dab4bfdc9f3": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_a0d0b28b1bed45c09594979f477cd0c6", "style": "IPY_MODEL_09e775321e064dc186a94213be9cd428", "value": "Load SNP genotypes: 100%"}}, "80d5d938c8354978b61b105a77b21962": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_a4e4b43dfba64cabbb78a9dbac0755f1", "IPY_MODEL_8ac3cd931f444ca2beb8f039ff71baa4", "IPY_MODEL_b500c0b0f2104106acbb8ca407d50af9"], "layout": "IPY_MODEL_eb8a7430fb4e44c2bf10ea308bab6c45"}}, "892915945c434ea287d4a28c491f8cb9": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "8ac3cd931f444ca2beb8f039ff71baa4": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_d13cfbe6a5784352bc1cbd42be107bbc", "max": 9675, "style": "IPY_MODEL_94fec11344074e9dab9e96d0950f9461", "value": 9675}}, "8c53d25142c040c0b9c0b09c84fa6a24": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_597023dfd6b64751bf548422d50f4ea9", "max": 6, "style": "IPY_MODEL_ecdb225ff8cb4e159b8982819eadee9b", "value": 6}}, "8cd22ebdfff5403baab477a4eb37bce6": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "9303af7e9f644f15bae779b48e0a9f5e": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_9fa070c8979e447b861b8627833a3dd9", "style": "IPY_MODEL_2ae68bf5b37343c185d84add35e8818f", "value": "Compute SNP effects: 100%"}}, "93d28dc09e80429b949589070a80db84": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "94fec11344074e9dab9e96d0950f9461": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "962547f642b14d63bb2f01d4750ad3bf": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_1abbaaba9cf044c6b95d3ce434b21ef1", "style": "IPY_MODEL_b6a63bb6c5704d1c8ad30620fa5ec863", "value": "Compute allele frequencies: 100%"}}, "9a2cc242763b45099db10e53e0552626": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "9b21e6cc977c414a99f9faa3421668ba": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_9a2cc242763b45099db10e53e0552626", "style": "IPY_MODEL_c6ec0a29215c40f882a178675b25479f", "value": " 32/32 [00:00&lt;00:00, 75.38it/s]"}}, "9bf1eb790a1d4766958ffe280efbde24": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "9eacb59f381d4a0f89505f67a93475c6": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_198865cffd8a4e23a532c729e1704615", "style": "IPY_MODEL_f966aec30a114ee3a24a0da63a9f0136", "value": " 6/6 [00:00&lt;00:00, 60.16it/s]"}}, "9fa070c8979e447b861b8627833a3dd9": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "a037168605524e8e957941a1cf28650e": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_24a27df073604a0d82272c87326ed191", "max": 32, "style": "IPY_MODEL_466ca83f6c324f30ae3ca3275c3ab78b", "value": 32}}, "a068f56c16fa48d1a216b328401ffa05": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "a0d0b28b1bed45c09594979f477cd0c6": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "a0f38c1810ab4229a9bd25d23783514c": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "a27c69704ba84eaeb8b11371978be68f": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "a2a9769ec4e34a6592c02caa551e6184": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_67d78a35a5724ca086ef51d8856cb74d", "style": "IPY_MODEL_be388a1368f846fd95baa721ac10c57b", "value": "Compute allele frequencies: 100%"}}, "a2b8d6ef5a064619ab9f0b9e7ecb23d8": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "a2bb0c4682ba4bbf8723bd1fc9e7dbdb": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "a42d8fc38b0e45c397757cfaacc129bb": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_2718927363974000af1ed9d7ddcdf973", "max": 9675, "style": "IPY_MODEL_bc06570d8cc14ef9866d41e2ebb37410", "value": 9675}}, "a4e4b43dfba64cabbb78a9dbac0755f1": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_1d82e912a75243a7a95085e4d1b013bd", "style": "IPY_MODEL_721d20b3c62343b793c9f60fdec870e2", "value": "Compute SNP effects: 100%"}}, "a9b565de86164bd8be90e5943132a60a": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "b1046cf56ac04bf6be724de035707363": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_9303af7e9f644f15bae779b48e0a9f5e", "IPY_MODEL_e1def5ea18014c38abb8c234137645f7", "IPY_MODEL_270959f2f69f4db1bcdf2e885c5e8995"], "layout": "IPY_MODEL_4836d38622be48aab4e461f4e0c35a53"}}, "b32c6f3983744d7fa734d779d0ee1ec5": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "b388cc197dd64460bce3906111ce643e": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_a9b565de86164bd8be90e5943132a60a", "style": "IPY_MODEL_f8f223cb9ad94ed185fb1a589fa56f07", "value": "Compute allele frequencies: 100%"}}, "b4343a439bd840f493f5db9c291bf327": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "b48f986ed5bd4d04a932a42ea1fe3cbe": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_615441c5e6704292b65eaba257b563f1", "IPY_MODEL_8c53d25142c040c0b9c0b09c84fa6a24", "IPY_MODEL_355560fa88c2458f90079129ea57df2b"], "layout": "IPY_MODEL_c3c58da7052743e086de9152d6349b55"}}, "b500c0b0f2104106acbb8ca407d50af9": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_01feec3ce0a44d8d8dcefbd38adec76d", "style": "IPY_MODEL_d30577fdb2cd4e9d88431aaa8ff9d6a9", "value": " 9675/9675 [00:00&lt;00:00,  1.85it/s]"}}, "b574e1bc044043ac9ca380b8863df5b7": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_22da57a2a1754af39446090815ac5218", "style": "IPY_MODEL_b32c6f3983744d7fa734d779d0ee1ec5", "value": " 9675/9675 [00:00&lt;00:00, 28590.98it/s]"}}, "b5e3dd8ae5544641b7e47825eb3ab086": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "b6a63bb6c5704d1c8ad30620fa5ec863": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "b78ddc81e8c14c70b6c59cd216703779": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "b7c48493b45445a1bcd885bfd59004e2": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_c69c5b32d55545efbf4120e1580d2648", "style": "IPY_MODEL_f5bebc45ef76487992b1f82fa999ed86", "value": " 32/32 [00:00&lt;00:00, 57.48it/s]"}}, "ba66918cf3bd42fb8a7e481b3ed5e5e0": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "ba69408b986a4483b215c75086a3ea22": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "bc06570d8cc14ef9866d41e2ebb37410": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "bcda8a6537834de1befff12992a89159": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "bceb261c28dc46c6aac5c68b3836b22e": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "be388a1368f846fd95baa721ac10c57b": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "bf4667ea7d9844238969ae84119bd992": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_a2a9769ec4e34a6592c02caa551e6184", "IPY_MODEL_4d178fc793fd4c67a71e1e90012e017b", "IPY_MODEL_c350cf57e2a54053b6ca3ba3ad871355"], "layout": "IPY_MODEL_9bf1eb790a1d4766958ffe280efbde24"}}, "c0bfd4c9151e4abd8d1179904ec2dfc9": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "c350cf57e2a54053b6ca3ba3ad871355": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_a2bb0c4682ba4bbf8723bd1fc9e7dbdb", "style": "IPY_MODEL_31a55dab84e34de7ae58647ecdf0d53c", "value": " 6/6 [00:00&lt;00:00, 59.65it/s]"}}, "c3c58da7052743e086de9152d6349b55": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "c53c9fc1eeae47cdafae586f29788678": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "c69c5b32d55545efbf4120e1580d2648": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "c6ec0a29215c40f882a178675b25479f": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "ca7e9a942f1a4fa797b8f56215a1de42": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "cb0ca8b8dad841dba4af8c3dd8229631": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_c53c9fc1eeae47cdafae586f29788678", "max": 6, "style": "IPY_MODEL_892915945c434ea287d4a28c491f8cb9", "value": 6}}, "cf718db6796740f0b94f1063346fdce1": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "cf7e0dab672243f7afe4c3c47c2ddd9d": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "d0cb7f9464384a05ac34de9da43a27e1": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "d13cfbe6a5784352bc1cbd42be107bbc": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "d165dc10dd004f7088c2d8b9aa0625d7": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "d30577fdb2cd4e9d88431aaa8ff9d6a9": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "d6dbfdbd0b09460799b4456ec57e7127": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "db04a6716a6a4caa8ac70513df91318e": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_b388cc197dd64460bce3906111ce643e", "IPY_MODEL_cb0ca8b8dad841dba4af8c3dd8229631", "IPY_MODEL_ee4d0ab9d77748a2b0645ff2b3e63d51"], "layout": "IPY_MODEL_dfee9ee48f9a43628fef9bcd0582c7d7"}}, "dfee9ee48f9a43628fef9bcd0582c7d7": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "e1def5ea18014c38abb8c234137645f7": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_a27c69704ba84eaeb8b11371978be68f", "max": 9675, "style": "IPY_MODEL_65aa547efcf1478dbf2840c6865af214", "value": 9675}}, "e2b9df016be346a8a0ec7a064f7874a4": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "e30366f21178417fa4cc6e972f9f5095": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "e4603bc7527c42e58eacc29ea1930df8": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "e4e478deb89946acaaf7522258156cc6": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_17d5610bc5744ac78a017248e9024b2a", "IPY_MODEL_6cb7d773758d42fc9c1c6bc131b154b8", "IPY_MODEL_e682e229d4994c7981ce6de36db4d850"], "layout": "IPY_MODEL_2108a6c2f0e8473f9465050d08f9278c"}}, "e61980ac81f4419e95594c8414ce3130": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_695d241857564936a6e07dcd9007860c", "style": "IPY_MODEL_ca7e9a942f1a4fa797b8f56215a1de42", "value": " 9675/9675 [00:00&lt;00:00, 22987.98it/s]"}}, "e682e229d4994c7981ce6de36db4d850": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_c0bfd4c9151e4abd8d1179904ec2dfc9", "style": "IPY_MODEL_e30366f21178417fa4cc6e972f9f5095", "value": " 32/32 [00:00&lt;00:00, 40.60it/s]"}}, "e72011bffa6347f2a9cb820cfd611adf": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "e8869981c9cf48a0b9ddaae87e416703": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HBoxModel", "state": {"children": ["IPY_MODEL_50de8fd4b8c34326a58de2c550a66048", "IPY_MODEL_4f75f60e4a0f43ca97fa70a239aed625", "IPY_MODEL_e61980ac81f4419e95594c8414ce3130"], "layout": "IPY_MODEL_35a44715966543a580aca8be0b1a43d9"}}, "e8e8c21b4a96459a9801835c02cea33b": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "ea1469ca85594801a87dab85308ca6ae": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "eb5c25ba5b5f42068ef12526051ba499": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "eb8a7430fb4e44c2bf10ea308bab6c45": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "ecdb225ff8cb4e159b8982819eadee9b": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "ee4d0ab9d77748a2b0645ff2b3e63d51": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "HTMLModel", "state": {"layout": "IPY_MODEL_a068f56c16fa48d1a216b328401ffa05", "style": "IPY_MODEL_570dc7cd94bd4c308b5b80ffef8276f1", "value": " 6/6 [00:00&lt;00:00, 53.27it/s]"}}, "efea2c9de56044148caab1cd7fc8018d": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "ProgressStyleModel", "state": {"description_width": ""}}, "f05d35c7b9284d8397b99c430c0284b2": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "FloatProgressModel", "state": {"bar_style": "success", "layout": "IPY_MODEL_d0cb7f9464384a05ac34de9da43a27e1", "max": 6, "style": "IPY_MODEL_4503cc6920354eeca0f6a0b1257a1b05", "value": 6}}, "f1574ba01610458a8876b76f8f59f807": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "f46eed124c8740e8b5dfea671fa05a44": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "f4d97bd502974cd7b455befa930858ea": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "f5bebc45ef76487992b1f82fa999ed86": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "f765d760859d410eacc2c7c0c567aac1": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "f8f223cb9ad94ed185fb1a589fa56f07": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "f966aec30a114ee3a24a0da63a9f0136": {"model_module": "@jupyter-widgets/controls", "model_module_version": "1.5.0", "model_name": "DescriptionStyleModel", "state": {"description_width": ""}}, "f9dcd46db93b457784b840fa0def44a2": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}, "f9e452b4cc654a499daa5b729ad45806": {"model_module": "@jupyter-widgets/base", "model_module_version": "1.2.0", "model_name": "LayoutModel", "state": {}}}, "version_major": 2, "version_minor": 0}</script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script crossorigin="anonymous" data-jupyter-widgets-cdn="https://cdn.jsdelivr.net/npm/" src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@1.0.6/dist/embed-amd.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'workshop-2/module-1-pandas';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Module 2 - The genetic basis of metabolic resistance to insecticides in Anopheles mosquitoes" href="module-2-metabolic-ir.html" />
    <link rel="prev" title="Workshop 2 - Metabolic insecticide resistance and copy number variation" href="about.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../home.html">
  
  
  
  
  
    <p class="title logo__title">Training course in data analysis for genomic surveillance of African malaria vectors</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../home.html">
                    Home
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 1</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-1/about.html">Workshop 1 - Pyrethroid target-site resistance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-1/module-1-colab.html">Module 1 - Interactive cloud computing with Google Colaboratory</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-1/module-2-sample-metadata.html">Module 2 - Accessing and exploring <em>Anopheles</em> genomic data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-1/module-3-genome.html">Module 3 - Navigating the <em>Anopheles gambiae</em> reference genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-1/module-4-vgsc-snps.html">Module 4 - Analysing SNPs in the <em>Vgsc</em> gene</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-1/journal-club.html">Journal club - Clarkson et al. (2021)</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 2</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="about.html">Workshop 2 - Metabolic insecticide resistance and copy number variation</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Module 1 - Introduction to pandas DataFrames</a></li>
<li class="toctree-l1"><a class="reference internal" href="module-2-metabolic-ir.html">Module 2 - The genetic basis of metabolic resistance to insecticides in <em>Anopheles</em> mosquitoes</a></li>
<li class="toctree-l1"><a class="reference internal" href="module-3-cnv-calling.html">Module 3 - Detecting CNVs from genome sequence data</a></li>
<li class="toctree-l1"><a class="reference internal" href="module-4-cnv-frequencies.html">Module 4 - Analysing CNV frequencies at metabolic resistance genes</a></li>
<li class="toctree-l1"><a class="reference internal" href="journal-club.html">Journal club - Lucas et al. (2019)</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 3</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-3/about.html">Workshop 3 - Population structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-3/module-1-plotly.html">Module 1 - Plotting with Plotly Express</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-3/module-2-population-structure.html">Module 2 - The biology of population structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-3/module-3-snps.html">Module 3 - Genome-wide SNP data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-3/module-4-pca.html">Module 4 - Detecting population structure using PCA</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-3/journal-club.html">Journal club - Ag1000G Consortium (2017)</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 4</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-4/about.html">Workshop 4 - Cryptic species</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-4/module-1-numpy.html">Module 1 - NumPy arrays</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-4/module-2-cryptic-species.html">Module 2 - Species complexes and cryptic taxa in malaria vectors</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-4/module-3-aims.html">Module 3 - Ancestry-informative markers (AIMs)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-4/module-4-taxon-analysis.html">Module 4 - Discovering cryptic taxa</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-4/journal-club.html">Journal club - Tennessen et al. (2020)</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 5</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-5/about.html">Workshop 5 - Genetic diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-5/module-1-xarray.html">Module 1 - Xarray datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-5/module-2-diversity.html">Module 2 - Introduction to genetic diversity</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-5/module-3-stats.html">Module 3 - Genetic diversity summary statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-5/module-4-roh.html">Module 4 - Heterozygosity and runs of homozygosity</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 6</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-6/about.html">Workshop 6 - Detecting genes under recent positive selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-6/module-1-haplotypes.html">Module 1 - Haplotypes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-6/module-2-selection.html">Module 2 - The biology of selective sweeps</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-6/module-3-gwss.html">Module 3 - Detecting new forms of insecticide resistance using selection scans</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-6/module-4-primers.html">Module 4 - AgamPrimer - Primer Design in <em>Anopheles gambiae</em></a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 7</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-7/about.html">Workshop 7 - Gene flow and the spread of insecticide resistance</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-7/module-1-op-ir.html">Module 1 - Organophosphate resistance markers (combining SNP and CNV data)</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-7/module-2-gene-flow.html">Module 2 - Adaptive gene flow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-7/module-3-haplotype-clustering.html">Module 3 - Investigating gene flow at loci of interest using haplotype clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-7/module-4-haplotype-networks.html">Module 4 - Investigating gene flow at loci of interest using haplotype networks</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Workshop 8</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../workshop-8/about.html">Workshop 8 - Analysis planning and reporting</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-8/module-1-planning-analysis.html">Module 1 - Planning a genomic surveillance exploratory data analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-8/module-2-presenting-results.html">Module 2 - Presenting results of a genomic surveillance data analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-8/module-3-genomic-surveillance-tanzania.html">Module 3 - Genomic surveillance of malaria vectors in Tanzania</a></li>
<li class="toctree-l1"><a class="reference internal" href="../workshop-8/module-4-genomic-surveillance-ghana.html">Module 4 - Genomic surveillance of malaria vectors in Ghana</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-launch-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Launch interactive content">
    <i class="fas fa-rocket"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://colab.research.google.com/github/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io/blob/master/docs/workshop-2/module-1-pandas.ipynb" target="_blank"
   class="btn btn-sm dropdown-item"
   title="Launch onColab"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  
    <img src="../_static/images/logo_colab.png">
  </span>
<span class="btn__text-container">Colab</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io" target="_blank"
   class="btn btn-sm btn-source-repository-button dropdown-item"
   title="Source repository"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="btn__text-container">Repository</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/anopheles-genomic-surveillance/anopheles-genomic-surveillance.github.io/issues/new?title=Issue%20on%20page%20%2Fworkshop-2/module-1-pandas.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/workshop-2/module-1-pandas.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Module 1 - Introduction to pandas DataFrames</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-objectives">Learning objectives</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#lecture">Lecture</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#english">English</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#francais">Franais</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup">Setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pandas-dataframes">pandas DataFrames</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-sample-metadata">Example: sample metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-genome-annotations">Example: genome annotations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-snp-allele-frequencies">Example: SNP allele frequencies</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#aside-variable-naming-styles">Aside: variable naming styles</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-reading-from-a-csv-file">Example: reading from a CSV file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-creating-a-dataframe-from-scratch">Example: creating a DataFrame from scratch</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#structure-of-a-dataframe">Structure of a DataFrame</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#counting-rows-and-columns">Counting rows and columns</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-columns">Accessing columns</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-column-names">Accessing column names</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-a-single-column">Accessing a single column</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#investigating-data-in-a-column">Investigating data in a column</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-multiple-columns">Accessing multiple columns</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-rows">Accessing rows</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-a-single-row">Accessing a single row</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-multiple-rows">Accessing multiple rows</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#indexes">Indexes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#queries">Queries</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input-output">Input/output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#topics-not-covered-here">Topics not covered here</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#further-reading">Further reading</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <p><img alt="banner" src="../_images/banner.jpg" /></p>
<p><em><strong><a class="reference internal" href="about.html"><span class="doc std std-doc">Workshop 2</span></a> - Training course in data analysis for genomic surveillance of African malaria vectors</strong></em></p>
<hr class="docutils" />
<section id="module-1-introduction-to-pandas-dataframes">
<h1>Module 1 - Introduction to pandas DataFrames<a class="headerlink" href="#module-1-introduction-to-pandas-dataframes" title="Permalink to this heading">#</a></h1>
<p><strong>Theme: Tools &amp; Technology</strong></p>
<p>In this module were going to take a closer look at a Python package called <a class="reference external" href="https://pandas.pydata.org/">pandas</a>, which is an essential part of our data analysis toolkit. We use pandas whenever we need to store, manipulate or analyse tables of data, also known as DataFrames. In this module well focus on using pandas to analyse genomic data, but it can be used to analyse any kind of data, so its a very useful tool to have at your fingertips.</p>
<section id="learning-objectives">
<h2>Learning objectives<a class="headerlink" href="#learning-objectives" title="Permalink to this heading">#</a></h2>
<p>After this module you will be able to:</p>
<ul class="simple">
<li><p>Explain what a DataFrame is</p></li>
<li><p>Access columns and rows from a DataFrame</p></li>
<li><p>Use indexes and queries</p></li>
<li><p>Write a DataFrame to a file</p></li>
</ul>
<p>Please note that this module will cover only a small part of the features that pandas provides - see the links at then end of this notebook if youd like to go further.</p>
</section>
<section id="lecture">
<h2>Lecture<a class="headerlink" href="#lecture" title="Permalink to this heading">#</a></h2>
<section id="english">
<h3>English<a class="headerlink" href="#english" title="Permalink to this heading">#</a></h3>
<div class="cell tag_remove-input docutils container">
<div class="cell_output docutils container">
<div class="output text_html"><iframe width="560" height="315" src="https://www.youtube.com/embed/iIYftkvS0eM" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
</div></div>
</div>
</section>
<section id="francais">
<h3>Franais<a class="headerlink" href="#francais" title="Permalink to this heading">#</a></h3>
<div class="cell tag_remove-input docutils container">
<div class="cell_output docutils container">
<div class="output text_html"><iframe width="560" height="315" src="https://www.youtube.com/embed/RdyXA31iHz8" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
</div></div>
</div>
<p>Please note that the code in the cells below might differ from that shown in the video. This can happen because Python packages and their dependencies change due to updates, necessitating tweaks to the code.</p>
</section>
</section>
<section id="setup">
<h2>Setup<a class="headerlink" href="#setup" title="Permalink to this heading">#</a></h2>
<p>Lets install the <code class="docutils literal notranslate"><span class="pre">malariagen_data</span></code> package so we have easy access to some genomic data to demonstrate with. Pandas is already installed on colab.</p>
<div class="cell tag_remove-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip<span class="w"> </span>install<span class="w"> </span>-q<span class="w"> </span>--no-warn-conflicts<span class="w"> </span>malariagen_data
</pre></div>
</div>
</div>
</div>
<p>Now import the <code class="docutils literal notranslate"><span class="pre">malariagen_data</span></code> and <code class="docutils literal notranslate"><span class="pre">pandas</span></code> packages. Note that by convention we use <code class="docutils literal notranslate"><span class="pre">pd</span></code> as an abbreviation for <code class="docutils literal notranslate"><span class="pre">pandas</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">malariagen_data</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
</pre></div>
</div>
</div>
</div>
<p>Set up access to the MalariaGEN Ag3.0 data resource.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ag3</span> <span class="o">=</span> <span class="n">malariagen_data</span><span class="o">.</span><span class="n">Ag3</span><span class="p">()</span>
<span class="n">ag3</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/javascript">(function(root) {
  function now() {
    return new Date();
  }

  const force = true;

  if (typeof root._bokeh_onload_callbacks === "undefined" || force === true) {
    root._bokeh_onload_callbacks = [];
    root._bokeh_is_loading = undefined;
  }

const JS_MIME_TYPE = 'application/javascript';
  const HTML_MIME_TYPE = 'text/html';
  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';
  const CLASS_NAME = 'output_bokeh rendered_html';

  /**
   * Render data to the DOM node
   */
  function render(props, node) {
    const script = document.createElement("script");
    node.appendChild(script);
  }

  /**
   * Handle when an output is cleared or removed
   */
  function handleClearOutput(event, handle) {
    const cell = handle.cell;

    const id = cell.output_area._bokeh_element_id;
    const server_id = cell.output_area._bokeh_server_id;
    // Clean up Bokeh references
    if (id != null && id in Bokeh.index) {
      Bokeh.index[id].model.document.clear();
      delete Bokeh.index[id];
    }

    if (server_id !== undefined) {
      // Clean up Bokeh references
      const cmd_clean = "from bokeh.io.state import curstate; print(curstate().uuid_to_server['" + server_id + "'].get_sessions()[0].document.roots[0]._id)";
      cell.notebook.kernel.execute(cmd_clean, {
        iopub: {
          output: function(msg) {
            const id = msg.content.text.trim();
            if (id in Bokeh.index) {
              Bokeh.index[id].model.document.clear();
              delete Bokeh.index[id];
            }
          }
        }
      });
      // Destroy server and session
      const cmd_destroy = "import bokeh.io.notebook as ion; ion.destroy_server('" + server_id + "')";
      cell.notebook.kernel.execute(cmd_destroy);
    }
  }

  /**
   * Handle when a new output is added
   */
  function handleAddOutput(event, handle) {
    const output_area = handle.output_area;
    const output = handle.output;

    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only
    if ((output.output_type != "display_data") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {
      return
    }

    const toinsert = output_area.element.find("." + CLASS_NAME.split(' ')[0]);

    if (output.metadata[EXEC_MIME_TYPE]["id"] !== undefined) {
      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];
      // store reference to embed id on output_area
      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE]["id"];
    }
    if (output.metadata[EXEC_MIME_TYPE]["server_id"] !== undefined) {
      const bk_div = document.createElement("div");
      bk_div.innerHTML = output.data[HTML_MIME_TYPE];
      const script_attrs = bk_div.children[0].attributes;
      for (let i = 0; i < script_attrs.length; i++) {
        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);
        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent
      }
      // store reference to server id on output_area
      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE]["server_id"];
    }
  }

  function register_renderer(events, OutputArea) {

    function append_mime(data, metadata, element) {
      // create a DOM node to render to
      const toinsert = this.create_output_subarea(
        metadata,
        CLASS_NAME,
        EXEC_MIME_TYPE
      );
      this.keyboard_manager.register_events(toinsert);
      // Render to node
      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};
      render(props, toinsert[toinsert.length - 1]);
      element.append(toinsert);
      return toinsert
    }

    /* Handle when an output is cleared or removed */
    events.on('clear_output.CodeCell', handleClearOutput);
    events.on('delete.Cell', handleClearOutput);

    /* Handle when a new output is added */
    events.on('output_added.OutputArea', handleAddOutput);

    /**
     * Register the mime type and append_mime function with output_area
     */
    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {
      /* Is output safe? */
      safe: true,
      /* Index of renderer in `output_area.display_order` */
      index: 0
    });
  }

  // register the mime type if in Jupyter Notebook environment and previously unregistered
  if (root.Jupyter !== undefined) {
    const events = require('base/js/events');
    const OutputArea = require('notebook/js/outputarea').OutputArea;

    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {
      register_renderer(events, OutputArea);
    }
  }
  if (typeof (root._bokeh_timeout) === "undefined" || force === true) {
    root._bokeh_timeout = Date.now() + 5000;
    root._bokeh_failed_load = false;
  }

  const NB_LOAD_WARNING = {'data': {'text/html':
     "<div style='background-color: #fdd'>\n"+
     "<p>\n"+
     "BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \n"+
     "may be due to a slow or bad network connection. Possible fixes:\n"+
     "</p>\n"+
     "<ul>\n"+
     "<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\n"+
     "<li>use INLINE resources instead, as so:</li>\n"+
     "</ul>\n"+
     "<code>\n"+
     "from bokeh.resources import INLINE\n"+
     "output_notebook(resources=INLINE)\n"+
     "</code>\n"+
     "</div>"}};

  function display_loaded() {
    const el = document.getElementById(null);
    if (el != null) {
      el.textContent = "BokehJS is loading...";
    }
    if (root.Bokeh !== undefined) {
      if (el != null) {
        el.textContent = "BokehJS " + root.Bokeh.version + " successfully loaded.";
      }
    } else if (Date.now() < root._bokeh_timeout) {
      setTimeout(display_loaded, 100)
    }
  }

  function run_callbacks() {
    try {
      root._bokeh_onload_callbacks.forEach(function(callback) {
        if (callback != null)
          callback();
      });
    } finally {
      delete root._bokeh_onload_callbacks
    }
    console.debug("Bokeh: all callbacks have finished");
  }

  function load_libs(css_urls, js_urls, callback) {
    if (css_urls == null) css_urls = [];
    if (js_urls == null) js_urls = [];

    root._bokeh_onload_callbacks.push(callback);
    if (root._bokeh_is_loading > 0) {
      console.debug("Bokeh: BokehJS is being loaded, scheduling callback at", now());
      return null;
    }
    if (js_urls == null || js_urls.length === 0) {
      run_callbacks();
      return null;
    }
    console.debug("Bokeh: BokehJS not loaded, scheduling load and callback at", now());
    root._bokeh_is_loading = css_urls.length + js_urls.length;

    function on_load() {
      root._bokeh_is_loading--;
      if (root._bokeh_is_loading === 0) {
        console.debug("Bokeh: all BokehJS libraries/stylesheets loaded");
        run_callbacks()
      }
    }

    function on_error(url) {
      console.error("failed to load " + url);
    }

    for (let i = 0; i < css_urls.length; i++) {
      const url = css_urls[i];
      const element = document.createElement("link");
      element.onload = on_load;
      element.onerror = on_error.bind(null, url);
      element.rel = "stylesheet";
      element.type = "text/css";
      element.href = url;
      console.debug("Bokeh: injecting link tag for BokehJS stylesheet: ", url);
      document.body.appendChild(element);
    }

    for (let i = 0; i < js_urls.length; i++) {
      const url = js_urls[i];
      const element = document.createElement('script');
      element.onload = on_load;
      element.onerror = on_error.bind(null, url);
      element.async = false;
      element.src = url;
      console.debug("Bokeh: injecting script tag for BokehJS library: ", url);
      document.head.appendChild(element);
    }
  };

  function inject_raw_css(css) {
    const element = document.createElement("style");
    element.appendChild(document.createTextNode(css));
    document.body.appendChild(element);
  }

  const js_urls = ["https://cdn.bokeh.org/bokeh/release/bokeh-3.1.1.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.1.1.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.1.1.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.1.1.min.js", "https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.1.1.min.js"];
  const css_urls = [];

  const inline_js = [    function(Bokeh) {
      Bokeh.set_log_level("info");
    },
function(Bokeh) {
    }
  ];

  function run_inline_js() {
    if (root.Bokeh !== undefined || force === true) {
          for (let i = 0; i < inline_js.length; i++) {
      inline_js[i].call(root, root.Bokeh);
    }
} else if (Date.now() < root._bokeh_timeout) {
      setTimeout(run_inline_js, 100);
    } else if (!root._bokeh_failed_load) {
      console.log("Bokeh: BokehJS failed to load within specified timeout.");
      root._bokeh_failed_load = true;
    } else if (force !== true) {
      const cell = $(document.getElementById(null)).parents('.cell').data().cell;
      cell.output_area.append_execute_result(NB_LOAD_WARNING)
    }
  }

  if (root._bokeh_is_loading === 0) {
    console.debug("Bokeh: BokehJS loaded, going straight to plotting");
    run_inline_js();
  } else {
    load_libs(css_urls, js_urls, function() {
      console.debug("Bokeh: BokehJS plotting callback run at", now());
      run_inline_js();
    });
  }
}(window));</script><div class="output text_html">
            <table class="malariagen-ag3">
                <thead>
                    <tr>
                        <th style="text-align: left" colspan="2">MalariaGEN Ag3 API client</th>
                    </tr>
                    <tr><td colspan="2" style="text-align: left">
                        Please note that data are subject to terms of use,
                        for more information see <a href="https://www.malariagen.net/data">
                        the MalariaGEN website</a> or contact data@malariagen.net.
                        See also the <a href="https://malariagen.github.io/vector-data/ag3/api.html">Ag3 API docs</a>.
                    </td></tr>
                </thead>
                <tbody>
                    <tr>
                        <th style="text-align: left">
                            Storage URL
                        </th>
                        <td>gs://vo_agam_release/</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            Data releases available
                        </th>
                        <td>3.0</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            Results cache
                        </th>
                        <td>None</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            Cohorts analysis
                        </th>
                        <td>20230516</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            AIM analysis
                        </th>
                        <td>20220528</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            Site filters analysis
                        </th>
                        <td>dt_20200416</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            Software version
                        </th>
                        <td>malariagen_data 7.10.0</td>
                    </tr>
                    <tr>
                        <th style="text-align: left">
                            Client location
                        </th>
                        <td>unknown</td>
                    </tr>
                </tbody>
            </table>
        </div></div>
</div>
</section>
<section id="pandas-dataframes">
<h2>pandas DataFrames<a class="headerlink" href="#pandas-dataframes" title="Permalink to this heading">#</a></h2>
<p>We use the pandas <code class="docutils literal notranslate"><span class="pre">DataFrame</span></code> class whenever we want to store and analyse data that can be structured as a table with rows and columns. Lets take a look at some examples of DataFrames that we met in the <a class="reference external" href="https://anopheles-genomic-surveillance.github.io/workshop-1/about.html">first workshop</a> of this training course.</p>
<section id="example-sample-metadata">
<h3>Example: sample metadata<a class="headerlink" href="#example-sample-metadata" title="Permalink to this heading">#</a></h3>
<p>In <a class="reference external" href="https://anopheles-genomic-surveillance.github.io/workshop-1/module-2-sample-metadata.html">workshop 1, module 2</a> we encountered the <a class="reference external" href="https://malariagen.github.io/vector-data/ag3/api.html#sample-metadata">sample_metadata()</a> function. This function returns a DataFrame of sample metadata. Lets call this function and assign the result to the variable <code class="docutils literal notranslate"><span class="pre">df_samples</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_samples</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">sample_metadata</span><span class="p">(</span><span class="n">sample_sets</span><span class="o">=</span><span class="s2">&quot;3.0&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Check what type of object we have received.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">type</span><span class="p">(</span><span class="n">df_samples</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>pandas.core.frame.DataFrame
</pre></div>
</div>
</div>
</div>
<p>Inspect the data in the DataFrame by running a code cell containing just the name of the variable.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_samples</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample_id</th>
      <th>partner_sample_id</th>
      <th>contributor</th>
      <th>country</th>
      <th>location</th>
      <th>year</th>
      <th>month</th>
      <th>latitude</th>
      <th>longitude</th>
      <th>sex_call</th>
      <th>...</th>
      <th>admin1_name</th>
      <th>admin1_iso</th>
      <th>admin2_name</th>
      <th>taxon</th>
      <th>cohort_admin1_year</th>
      <th>cohort_admin1_month</th>
      <th>cohort_admin1_quarter</th>
      <th>cohort_admin2_year</th>
      <th>cohort_admin2_month</th>
      <th>cohort_admin2_quarter</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AR0047-C</td>
      <td>LUA047</td>
      <td>Joao Pinto</td>
      <td>Angola</td>
      <td>Luanda</td>
      <td>2009</td>
      <td>4</td>
      <td>-8.884</td>
      <td>13.302</td>
      <td>F</td>
      <td>...</td>
      <td>Luanda</td>
      <td>AO-LUA</td>
      <td>Luanda</td>
      <td>coluzzii</td>
      <td>AO-LUA_colu_2009</td>
      <td>AO-LUA_colu_2009_04</td>
      <td>AO-LUA_colu_2009_Q2</td>
      <td>AO-LUA_Luanda_colu_2009</td>
      <td>AO-LUA_Luanda_colu_2009_04</td>
      <td>AO-LUA_Luanda_colu_2009_Q2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>AR0049-C</td>
      <td>LUA049</td>
      <td>Joao Pinto</td>
      <td>Angola</td>
      <td>Luanda</td>
      <td>2009</td>
      <td>4</td>
      <td>-8.884</td>
      <td>13.302</td>
      <td>F</td>
      <td>...</td>
      <td>Luanda</td>
      <td>AO-LUA</td>
      <td>Luanda</td>
      <td>coluzzii</td>
      <td>AO-LUA_colu_2009</td>
      <td>AO-LUA_colu_2009_04</td>
      <td>AO-LUA_colu_2009_Q2</td>
      <td>AO-LUA_Luanda_colu_2009</td>
      <td>AO-LUA_Luanda_colu_2009_04</td>
      <td>AO-LUA_Luanda_colu_2009_Q2</td>
    </tr>
    <tr>
      <th>2</th>
      <td>AR0051-C</td>
      <td>LUA051</td>
      <td>Joao Pinto</td>
      <td>Angola</td>
      <td>Luanda</td>
      <td>2009</td>
      <td>4</td>
      <td>-8.884</td>
      <td>13.302</td>
      <td>F</td>
      <td>...</td>
      <td>Luanda</td>
      <td>AO-LUA</td>
      <td>Luanda</td>
      <td>coluzzii</td>
      <td>AO-LUA_colu_2009</td>
      <td>AO-LUA_colu_2009_04</td>
      <td>AO-LUA_colu_2009_Q2</td>
      <td>AO-LUA_Luanda_colu_2009</td>
      <td>AO-LUA_Luanda_colu_2009_04</td>
      <td>AO-LUA_Luanda_colu_2009_Q2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>AR0061-C</td>
      <td>LUA061</td>
      <td>Joao Pinto</td>
      <td>Angola</td>
      <td>Luanda</td>
      <td>2009</td>
      <td>4</td>
      <td>-8.884</td>
      <td>13.302</td>
      <td>F</td>
      <td>...</td>
      <td>Luanda</td>
      <td>AO-LUA</td>
      <td>Luanda</td>
      <td>coluzzii</td>
      <td>AO-LUA_colu_2009</td>
      <td>AO-LUA_colu_2009_04</td>
      <td>AO-LUA_colu_2009_Q2</td>
      <td>AO-LUA_Luanda_colu_2009</td>
      <td>AO-LUA_Luanda_colu_2009_04</td>
      <td>AO-LUA_Luanda_colu_2009_Q2</td>
    </tr>
    <tr>
      <th>4</th>
      <td>AR0078-C</td>
      <td>LUA078</td>
      <td>Joao Pinto</td>
      <td>Angola</td>
      <td>Luanda</td>
      <td>2009</td>
      <td>4</td>
      <td>-8.884</td>
      <td>13.302</td>
      <td>F</td>
      <td>...</td>
      <td>Luanda</td>
      <td>AO-LUA</td>
      <td>Luanda</td>
      <td>coluzzii</td>
      <td>AO-LUA_colu_2009</td>
      <td>AO-LUA_colu_2009_04</td>
      <td>AO-LUA_colu_2009_Q2</td>
      <td>AO-LUA_Luanda_colu_2009</td>
      <td>AO-LUA_Luanda_colu_2009_04</td>
      <td>AO-LUA_Luanda_colu_2009_Q2</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>3076</th>
      <td>AD0494-C</td>
      <td>80-2-o-16</td>
      <td>Martin Donnelly</td>
      <td>Lab Cross</td>
      <td>LSTM</td>
      <td>-1</td>
      <td>-1</td>
      <td>53.409</td>
      <td>-2.969</td>
      <td>F</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>unassigned</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3077</th>
      <td>AD0495-C</td>
      <td>80-2-o-17</td>
      <td>Martin Donnelly</td>
      <td>Lab Cross</td>
      <td>LSTM</td>
      <td>-1</td>
      <td>-1</td>
      <td>53.409</td>
      <td>-2.969</td>
      <td>M</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>unassigned</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3078</th>
      <td>AD0496-C</td>
      <td>80-2-o-18</td>
      <td>Martin Donnelly</td>
      <td>Lab Cross</td>
      <td>LSTM</td>
      <td>-1</td>
      <td>-1</td>
      <td>53.409</td>
      <td>-2.969</td>
      <td>M</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>unassigned</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3079</th>
      <td>AD0497-C</td>
      <td>80-2-o-19</td>
      <td>Martin Donnelly</td>
      <td>Lab Cross</td>
      <td>LSTM</td>
      <td>-1</td>
      <td>-1</td>
      <td>53.409</td>
      <td>-2.969</td>
      <td>F</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>unassigned</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3080</th>
      <td>AD0498-C</td>
      <td>80-2-o-20</td>
      <td>Martin Donnelly</td>
      <td>Lab Cross</td>
      <td>LSTM</td>
      <td>-1</td>
      <td>-1</td>
      <td>53.409</td>
      <td>-2.969</td>
      <td>M</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>unassigned</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>3081 rows  30 columns</p>
</div></div></div>
</div>
<p>There are 3,081 rows and 26 columns in this DataFrame. Each row provides data about a single mosquito DNA sample that has been sequenced by MalariaGEN. The columns provide data about those mosquitoes, such as the country and year in which the mosquito was originally collected.</p>
<p>Note that the representation of the DataFrame above only shows some of the data - there are too many rows and columns to show all at once. The ellipsis character () shows where some data have been hidden.</p>
</section>
<section id="example-genome-annotations">
<h3>Example: genome annotations<a class="headerlink" href="#example-genome-annotations" title="Permalink to this heading">#</a></h3>
<p>In <a class="reference external" href="https://anopheles-genomic-surveillance.github.io/workshop-1/module-3-genome.html">workshop 1, module 3</a> we encountered the <a class="reference external" href="https://malariagen.github.io/vector-data/ag3/api.html#geneset">geneset()</a> function. This function returns a DataFrame of genome annotations. Lets call the function and assign the result to the variable <code class="docutils literal notranslate"><span class="pre">df_geneset</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_geneset</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">geneset</span><span class="p">()</span>
<span class="n">df_geneset</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>contig</th>
      <th>source</th>
      <th>type</th>
      <th>start</th>
      <th>end</th>
      <th>score</th>
      <th>strand</th>
      <th>phase</th>
      <th>ID</th>
      <th>Parent</th>
      <th>Name</th>
      <th>description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>2L</td>
      <td>VectorBase</td>
      <td>chromosome</td>
      <td>1</td>
      <td>49364325</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>2L</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2L</td>
      <td>VectorBase</td>
      <td>gene</td>
      <td>157348</td>
      <td>186936</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>AGAP004677</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>methylenetetrahydrofolate dehydrogenase(NAD ) ...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2L</td>
      <td>VectorBase</td>
      <td>mRNA</td>
      <td>157348</td>
      <td>181305</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>AGAP004677-RA</td>
      <td>AGAP004677</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>3</th>
      <td>2L</td>
      <td>VectorBase</td>
      <td>three_prime_UTR</td>
      <td>157348</td>
      <td>157495</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>AGAP004677-RA</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2L</td>
      <td>VectorBase</td>
      <td>exon</td>
      <td>157348</td>
      <td>157623</td>
      <td>NaN</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>AGAP004677-RA</td>
      <td>AGAP004677-RB-E4</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>196140</th>
      <td>Y_unplaced</td>
      <td>VectorBase</td>
      <td>five_prime_UTR</td>
      <td>47932</td>
      <td>48111</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>AGAP029375-RA</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>196141</th>
      <td>Y_unplaced</td>
      <td>VectorBase</td>
      <td>exon</td>
      <td>47932</td>
      <td>48138</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>AGAP029375-RA</td>
      <td>AGAP029375-RA-E2</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>196142</th>
      <td>Y_unplaced</td>
      <td>VectorBase</td>
      <td>CDS</td>
      <td>48112</td>
      <td>48138</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>AGAP029375-PA</td>
      <td>AGAP029375-RA</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>196143</th>
      <td>Y_unplaced</td>
      <td>VectorBase</td>
      <td>exon</td>
      <td>48301</td>
      <td>48385</td>
      <td>NaN</td>
      <td>+</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>AGAP029375-RA</td>
      <td>AGAP029375-RA-E3</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>196144</th>
      <td>Y_unplaced</td>
      <td>VectorBase</td>
      <td>CDS</td>
      <td>48301</td>
      <td>48385</td>
      <td>NaN</td>
      <td>+</td>
      <td>0.0</td>
      <td>AGAP029375-PA</td>
      <td>AGAP029375-RA</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
<p>196145 rows  12 columns</p>
</div></div></div>
</div>
<p>There are 196,145 rows and 12 columns in this DataFrame. Each row provides data about an annotated feature on the <em>Anopheles gambiae</em> AgamP4 reference genome, such as a gene. The columns provide data about those features, such as the contig (chromosome arm) on which the feature occurs, and the start and end coordinates.</p>
</section>
<section id="example-snp-allele-frequencies">
<h3>Example: SNP allele frequencies<a class="headerlink" href="#example-snp-allele-frequencies" title="Permalink to this heading">#</a></h3>
<p>In <a class="reference external" href="https://anopheles-genomic-surveillance.github.io/workshop-1/module-4-vgsc-snps.html">workshop 1, module 4</a>, we encountered the <a class="reference external" href="https://malariagen.github.io/vector-data/ag3/api.html#snp-allele-frequencies">snp_allele_frequencies()</a> function. This function returns a DataFrame of single nucleotide polymorphisms (SNPs) together with their allele frequencies in different mosquito cohorts. Lets call this function and assign the result to the variable <code class="docutils literal notranslate"><span class="pre">snp_freqs_df</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">snp_freqs_df</span> <span class="o">=</span> <span class="n">ag3</span><span class="o">.</span><span class="n">snp_allele_frequencies</span><span class="p">(</span>
    <span class="n">transcript</span><span class="o">=</span><span class="s2">&quot;AGAP004707-RD&quot;</span><span class="p">,</span> 
    <span class="n">cohorts</span><span class="o">=</span><span class="s2">&quot;admin1_year&quot;</span><span class="p">,</span> 
    <span class="n">sample_sets</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;AG1000G-BF-A&quot;</span><span class="p">,</span> <span class="s2">&quot;AG1000G-BF-B&quot;</span><span class="p">,</span> <span class="s2">&quot;AG1000G-BF-C&quot;</span><span class="p">],</span> 
<span class="p">)</span>
<span class="n">snp_freqs_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/vnd.jupyter.widget-view+json">{"model_id": "0a9bc466f51b4164b257102956ff9080", "version_major": 2, "version_minor": 0}</script><script type="application/vnd.jupyter.widget-view+json">{"model_id": "62f26b4b68dd461698800a2286704b56", "version_major": 2, "version_minor": 0}</script><script type="application/vnd.jupyter.widget-view+json">{"model_id": "e8869981c9cf48a0b9ddaae87e416703", "version_major": 2, "version_minor": 0}</script><div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th>pass_gamb_colu_arab</th>
      <th>pass_gamb_colu</th>
      <th>pass_arab</th>
      <th>frq_BF-07_gamb_2004</th>
      <th>frq_BF-09_colu_2012</th>
      <th>frq_BF-09_colu_2014</th>
      <th>frq_BF-09_gamb_2012</th>
      <th>frq_BF-09_gamb_2014</th>
      <th>max_af</th>
      <th>transcript</th>
      <th>effect</th>
      <th>impact</th>
      <th>ref_codon</th>
      <th>alt_codon</th>
      <th>aa_pos</th>
      <th>ref_aa</th>
      <th>alt_aa</th>
      <th>label</th>
    </tr>
    <tr>
      <th>contig</th>
      <th>position</th>
      <th>ref_allele</th>
      <th>alt_allele</th>
      <th>aa_change</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th rowspan="11" valign="top">2L</th>
      <th>2358252</th>
      <th>C</th>
      <th>T</th>
      <th>A32V</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.000000</td>
      <td>0.006098</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.006098</td>
      <td>AGAP004707-RD</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>gCa</td>
      <td>gTa</td>
      <td>32.0</td>
      <td>A</td>
      <td>V</td>
      <td>2L:2,358,252 C&gt;T (A32V)</td>
    </tr>
    <tr>
      <th>2358328</th>
      <th>T</th>
      <th>C</th>
      <th>NaN</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.000000</td>
      <td>0.006098</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.006098</td>
      <td>AGAP004707-RD</td>
      <td>INTRONIC</td>
      <td>MODIFIER</td>
      <td>None</td>
      <td>None</td>
      <td>NaN</td>
      <td>None</td>
      <td>None</td>
      <td>2L:2,358,328 T&gt;C</td>
    </tr>
    <tr>
      <th>2358441</th>
      <th>A</th>
      <th>T</th>
      <th>NaN</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.038462</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.038462</td>
      <td>AGAP004707-RD</td>
      <td>INTRONIC</td>
      <td>MODIFIER</td>
      <td>None</td>
      <td>None</td>
      <td>NaN</td>
      <td>None</td>
      <td>None</td>
      <td>2L:2,358,441 A&gt;T</td>
    </tr>
    <tr>
      <th>2358468</th>
      <th>A</th>
      <th>C</th>
      <th>NaN</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.269231</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.269231</td>
      <td>AGAP004707-RD</td>
      <td>INTRONIC</td>
      <td>MODIFIER</td>
      <td>None</td>
      <td>None</td>
      <td>NaN</td>
      <td>None</td>
      <td>None</td>
      <td>2L:2,358,468 A&gt;C</td>
    </tr>
    <tr>
      <th>2358581</th>
      <th>T</th>
      <th>A</th>
      <th>NaN</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.192308</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.192308</td>
      <td>AGAP004707-RD</td>
      <td>INTRONIC</td>
      <td>MODIFIER</td>
      <td>None</td>
      <td>None</td>
      <td>NaN</td>
      <td>None</td>
      <td>None</td>
      <td>2L:2,358,581 T&gt;A</td>
    </tr>
    <tr>
      <th>...</th>
      <th>...</th>
      <th>...</th>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>2431424</th>
      <th>G</th>
      <th>T</th>
      <th>G2055V</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.005051</td>
      <td>0.0</td>
      <td>0.005051</td>
      <td>AGAP004707-RD</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>gGc</td>
      <td>gTc</td>
      <td>2055.0</td>
      <td>G</td>
      <td>V</td>
      <td>2L:2,431,424 G&gt;T (G2055V)</td>
    </tr>
    <tr>
      <th>2431518</th>
      <th>C</th>
      <th>A</th>
      <th>A2086A</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.038462</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.038462</td>
      <td>AGAP004707-RD</td>
      <td>SYNONYMOUS_CODING</td>
      <td>LOW</td>
      <td>gcC</td>
      <td>gcA</td>
      <td>2086.0</td>
      <td>A</td>
      <td>A</td>
      <td>2L:2,431,518 C&gt;A (A2086A)</td>
    </tr>
    <tr>
      <th>2431527</th>
      <th>G</th>
      <th>C</th>
      <th>V2089V</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.038462</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.038462</td>
      <td>AGAP004707-RD</td>
      <td>SYNONYMOUS_CODING</td>
      <td>LOW</td>
      <td>gtG</td>
      <td>gtC</td>
      <td>2089.0</td>
      <td>V</td>
      <td>V</td>
      <td>2L:2,431,527 G&gt;C (V2089V)</td>
    </tr>
    <tr>
      <th>2431569</th>
      <th>C</th>
      <th>A</th>
      <th>V2103V</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.038462</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.038462</td>
      <td>AGAP004707-RD</td>
      <td>SYNONYMOUS_CODING</td>
      <td>LOW</td>
      <td>gtC</td>
      <td>gtA</td>
      <td>2103.0</td>
      <td>V</td>
      <td>V</td>
      <td>2L:2,431,569 C&gt;A (V2103V)</td>
    </tr>
    <tr>
      <th>2431613</th>
      <th>T</th>
      <th>C</th>
      <th>V2118A</th>
      <td>True</td>
      <td>True</td>
      <td>True</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.0</td>
      <td>0.005051</td>
      <td>0.0</td>
      <td>0.005051</td>
      <td>AGAP004707-RD</td>
      <td>NON_SYNONYMOUS_CODING</td>
      <td>MODERATE</td>
      <td>gTc</td>
      <td>gCc</td>
      <td>2118.0</td>
      <td>V</td>
      <td>A</td>
      <td>2L:2,431,613 T&gt;C (V2118A)</td>
    </tr>
  </tbody>
</table>
<p>9675 rows  18 columns</p>
</div></div></div>
</div>
<p>There are 9,675 rows and 18 columns in this DataFrame. Each row provides data about a SNP. The columns provide data about those SNPs, such as their effect, and the alternate allele frequency in a particular cohort of mosquitoes, e.g., the column named frq_BF-09_colu_2012 provides the frequency in <em>Anopheles coluzzii</em> mosquitoes collected in the BF-09 (Hauts-Bassins) administrative region of Burkina Faso in 2012.</p>
</section>
<section id="aside-variable-naming-styles">
<h3>Aside: variable naming styles<a class="headerlink" href="#aside-variable-naming-styles" title="Permalink to this heading">#</a></h3>
<p>You may have noticed that in the examples above, whenever we assigned a DataFrame to a variable, the name of the variable either started with the prefix <code class="docutils literal notranslate"><span class="pre">df_</span></code> or ended with the suffix <code class="docutils literal notranslate"><span class="pre">_df</span></code>. Here df stands for DataFrame, and is used just as a helpful reminder that the value of the variable is a pandas DataFrame. Using a prefix of suffix on the variable name like this is entirely optional and is up to your own coding style. You can name these variables whatever you like. Different instructors in this course prefer different styles, and you will see different styles being used in different modules.</p>
</section>
<section id="example-reading-from-a-csv-file">
<h3>Example: reading from a CSV file<a class="headerlink" href="#example-reading-from-a-csv-file" title="Permalink to this heading">#</a></h3>
<p>You can also create a pandas DataFrame by reading data from a file, such as a comma-separated value (CSV) file, or an Excel workbook file. Lets use the pandas <a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html">read_csv()</a> function to read data from a CSV file containing data from a US census that colab provides as example data for you to play with.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_housing</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;sample_data/california_housing_train.csv&quot;</span><span class="p">)</span>
<span class="n">df_housing</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>longitude</th>
      <th>latitude</th>
      <th>housing_median_age</th>
      <th>total_rooms</th>
      <th>total_bedrooms</th>
      <th>population</th>
      <th>households</th>
      <th>median_income</th>
      <th>median_house_value</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>-114.31</td>
      <td>34.19</td>
      <td>15.0</td>
      <td>5612.0</td>
      <td>1283.0</td>
      <td>1015.0</td>
      <td>472.0</td>
      <td>1.4936</td>
      <td>66900.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>-114.47</td>
      <td>34.40</td>
      <td>19.0</td>
      <td>7650.0</td>
      <td>1901.0</td>
      <td>1129.0</td>
      <td>463.0</td>
      <td>1.8200</td>
      <td>80100.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>-114.56</td>
      <td>33.69</td>
      <td>17.0</td>
      <td>720.0</td>
      <td>174.0</td>
      <td>333.0</td>
      <td>117.0</td>
      <td>1.6509</td>
      <td>85700.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>-114.57</td>
      <td>33.64</td>
      <td>14.0</td>
      <td>1501.0</td>
      <td>337.0</td>
      <td>515.0</td>
      <td>226.0</td>
      <td>3.1917</td>
      <td>73400.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>-114.57</td>
      <td>33.57</td>
      <td>20.0</td>
      <td>1454.0</td>
      <td>326.0</td>
      <td>624.0</td>
      <td>262.0</td>
      <td>1.9250</td>
      <td>65500.0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>16995</th>
      <td>-124.26</td>
      <td>40.58</td>
      <td>52.0</td>
      <td>2217.0</td>
      <td>394.0</td>
      <td>907.0</td>
      <td>369.0</td>
      <td>2.3571</td>
      <td>111400.0</td>
    </tr>
    <tr>
      <th>16996</th>
      <td>-124.27</td>
      <td>40.69</td>
      <td>36.0</td>
      <td>2349.0</td>
      <td>528.0</td>
      <td>1194.0</td>
      <td>465.0</td>
      <td>2.5179</td>
      <td>79000.0</td>
    </tr>
    <tr>
      <th>16997</th>
      <td>-124.30</td>
      <td>41.84</td>
      <td>17.0</td>
      <td>2677.0</td>
      <td>531.0</td>
      <td>1244.0</td>
      <td>456.0</td>
      <td>3.0313</td>
      <td>103600.0</td>
    </tr>
    <tr>
      <th>16998</th>
      <td>-124.30</td>
      <td>41.80</td>
      <td>19.0</td>
      <td>2672.0</td>
      <td>552.0</td>
      <td>1298.0</td>
      <td>478.0</td>
      <td>1.9797</td>
      <td>85800.0</td>
    </tr>
    <tr>
      <th>16999</th>
      <td>-124.35</td>
      <td>40.54</td>
      <td>52.0</td>
      <td>1820.0</td>
      <td>300.0</td>
      <td>806.0</td>
      <td>270.0</td>
      <td>3.0147</td>
      <td>94600.0</td>
    </tr>
  </tbody>
</table>
<p>17000 rows  9 columns</p>
</div></div></div>
</div>
<p>This DataFrame has 17,000 rows and 9 columns. Each row provides data about a block of housing. The columns provide data about the houses within a block, such as the total number of bedrooms and the number of households.</p>
<p>Note that you can upload your own files to Colab if you want to read in and analyse some of your own data with pandas.</p>
</section>
<section id="example-creating-a-dataframe-from-scratch">
<h3>Example: creating a DataFrame from scratch<a class="headerlink" href="#example-creating-a-dataframe-from-scratch" title="Permalink to this heading">#</a></h3>
<p>You can also create a DataFrame by providing the data within your code. To demonstrate this, lets create a DataFrame with data about characters in a famous science fiction book called The Hitchhikers Guide to the Galaxy.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_hitchhikers</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">{</span>
    <span class="s2">&quot;name&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;Arthur Dent&quot;</span><span class="p">,</span> <span class="s2">&quot;Tricia McMillan&quot;</span><span class="p">,</span> <span class="s2">&quot;Ford Prefect&quot;</span><span class="p">,</span> <span class="s2">&quot;Zaphod Beeblebrox&quot;</span><span class="p">,</span> <span class="s2">&quot;Marvin the Paranoid Android&quot;</span><span class="p">],</span>
    <span class="s2">&quot;taxon&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;human&quot;</span><span class="p">,</span> <span class="s2">&quot;human&quot;</span><span class="p">,</span> <span class="s2">&quot;betelgeusian&quot;</span><span class="p">,</span> <span class="s2">&quot;betelgeusian&quot;</span><span class="p">,</span> <span class="s2">&quot;robot&quot;</span><span class="p">],</span>
    <span class="s2">&quot;gender&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;male&quot;</span><span class="p">,</span> <span class="s2">&quot;female&quot;</span><span class="p">,</span> <span class="s2">&quot;male&quot;</span><span class="p">,</span> <span class="s2">&quot;male&quot;</span><span class="p">,</span> <span class="kc">None</span><span class="p">],</span>
    <span class="s2">&quot;number_of_heads&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
    <span class="s2">&quot;intelligence&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">100</span><span class="p">,</span> <span class="mi">180</span><span class="p">,</span> <span class="mi">120</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">5_000_000</span><span class="p">]</span>
<span class="p">})</span>
<span class="n">df_hitchhikers</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>taxon</th>
      <th>gender</th>
      <th>number_of_heads</th>
      <th>intelligence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Arthur Dent</td>
      <td>human</td>
      <td>male</td>
      <td>1</td>
      <td>100</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Tricia McMillan</td>
      <td>human</td>
      <td>female</td>
      <td>1</td>
      <td>180</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Ford Prefect</td>
      <td>betelgeusian</td>
      <td>male</td>
      <td>1</td>
      <td>120</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Zaphod Beeblebrox</td>
      <td>betelgeusian</td>
      <td>male</td>
      <td>2</td>
      <td>30</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Marvin the Paranoid Android</td>
      <td>robot</td>
      <td>None</td>
      <td>1</td>
      <td>5000000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>This DataFrame has 5 rows and 5 columns. Each row provides data about a fictional character in the book. The columns provide data about each character, such as their name, taxon and number of heads (some alien species can have more than one).</p>
</section>
</section>
<section id="structure-of-a-dataframe">
<h2>Structure of a DataFrame<a class="headerlink" href="#structure-of-a-dataframe" title="Permalink to this heading">#</a></h2>
<p>Hopefully this will already be obvious from the examples above, but a <a class="reference external" href="https://pandas.pydata.org/docs/user_guide/dsintro.html#dataframe">pandas DataFrame</a> is a data structure that is organised into <strong>rows</strong> and <strong>columns</strong>. Each DataFrame also has an <strong>index</strong>, which is a special column (or set of columns) that is shown in bold at the left of a DataFrame, and which well look at in more detail later.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%html</span>
<span class="p">&lt;</span><span class="nt">img</span> <span class="na">width</span><span class="o">=</span><span class="s">&quot;90%&quot;</span> <span class="na">height</span><span class="o">=</span><span class="s">&quot;90%&quot;</span><span class="na">src</span><span class="o">=</span><span class="s">&quot;https://storage.googleapis.com/vo_agam_release/reference/training-images/workshop-2/W2M1-1.PNG&quot;</span><span class="p">/&gt;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><img width="90%" height="90%"src="https://storage.googleapis.com/vo_agam_release/reference/training-images/workshop-2/W2M1-1.PNG"/>
</div></div>
</div>
</section>
<section id="counting-rows-and-columns">
<h2>Counting rows and columns<a class="headerlink" href="#counting-rows-and-columns" title="Permalink to this heading">#</a></h2>
<p>If a DataFrame is large, you may find it useful to know how many rows and columns it contains. The representation of a DataFrame shows at the bottom how many rows and columns are present in the DataFrame, but you can also use the built-in <code class="docutils literal notranslate"><span class="pre">len()</span></code> function if you want to access the number of rows or columns in your code.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">num_rows</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_hitchhikers</span><span class="p">)</span>
<span class="n">num_rows</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>5
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">num_cols</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df_hitchhikers</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<span class="n">num_cols</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>5
</pre></div>
</div>
</div>
</div>
<p><strong>Exercise 1 (English)</strong></p>
<ul class="simple">
<li><p>Count the number of rows and columns in the <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> DataFrame.</p></li>
</ul>
<p><strong>Exercice 1 (Franais)</strong></p>
<ul class="simple">
<li><p>Compter le nombre de lignes et de colonnes du DataFrame <code class="docutils literal notranslate"><span class="pre">df_samples</span></code>.</p></li>
</ul>
</section>
<section id="accessing-columns">
<h2>Accessing columns<a class="headerlink" href="#accessing-columns" title="Permalink to this heading">#</a></h2>
<p>Often you will have a DataFrame with many columns, and you many only need to work with data from a smaller number of columns, or just a single column. Accessing one or more columns from a DataFrame can be done using the square bracket syntax after the name of a DataFrame variable. Lets see how this works.</p>
<section id="accessing-column-names">
<h3>Accessing column names<a class="headerlink" href="#accessing-column-names" title="Permalink to this heading">#</a></h3>
<p>First, lets access the names of all the columns in a DataFrame, to see what columns are available. You can do this by accessing the <code class="docutils literal notranslate"><span class="pre">columns</span></code> property of a DataFrame. For simplicity, lets also convert the result to a Python list, using the built-in <code class="docutils literal notranslate"><span class="pre">list()</span></code> function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">col_names</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">df_hitchhikers</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
<span class="n">col_names</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;name&#39;, &#39;taxon&#39;, &#39;gender&#39;, &#39;number_of_heads&#39;, &#39;intelligence&#39;]
</pre></div>
</div>
</div>
</div>
<p>This can be useful especially if you have a lot of columns in a DataFrame.</p>
<p><strong>Exercise 2 (English)</strong></p>
<ul class="simple">
<li><p>Access the column names of the <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> DataFrame.</p></li>
</ul>
<p><strong>Exercice 2 (Franais)</strong></p>
<ul class="simple">
<li><p>Afficher les noms des colonnes du DataFrame <code class="docutils literal notranslate"><span class="pre">df_samples</span></code>.</p></li>
</ul>
</section>
<section id="accessing-a-single-column">
<h3>Accessing a single column<a class="headerlink" href="#accessing-a-single-column" title="Permalink to this heading">#</a></h3>
<p>To access a single column, give the column name as a string inside square brackets. Lets access the name column from the DataFrame of book characters, and assign the result to a new variable <code class="docutils literal notranslate"><span class="pre">character_name</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">character_name</span> <span class="o">=</span> <span class="n">df_hitchhikers</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>When you access a single column or row from a DataFrame, pandas returns a <a class="reference external" href="https://pandas.pydata.org/docs/user_guide/dsintro.html#series">Series</a> object to you. Pandas uses the Series class to hold data either from a single column or a single row, i.e., a one-dimensional sequence of data values.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">type</span><span class="p">(</span><span class="n">character_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>pandas.core.series.Series
</pre></div>
</div>
</div>
</div>
<p>We can also inspect the data in the column.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">character_name</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0                    Arthur Dent
1                Tricia McMillan
2                   Ford Prefect
3              Zaphod Beeblebrox
4    Marvin the Paranoid Android
Name: name, dtype: object
</pre></div>
</div>
</div>
</div>
</section>
<section id="investigating-data-in-a-column">
<h3>Investigating data in a column<a class="headerlink" href="#investigating-data-in-a-column" title="Permalink to this heading">#</a></h3>
<p>The pandas Series class has some useful functions for investigating the data in a DataFrame.</p>
<p>E.g., lets access the taxon column and use the <code class="docutils literal notranslate"><span class="pre">value_counts()</span></code> functions to get a summary of the categorical data in the column.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">character_taxon</span> <span class="o">=</span> <span class="n">df_hitchhikers</span><span class="p">[</span><span class="s2">&quot;taxon&quot;</span><span class="p">]</span>
<span class="n">character_taxon</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>taxon
human           2
betelgeusian    2
robot           1
Name: count, dtype: int64
</pre></div>
</div>
</div>
</div>
<p>The pandas Series class also has a <code class="docutils literal notranslate"><span class="pre">describe()</span></code> function which can be useful for investigating numerical data. Lets use it with the <code class="docutils literal notranslate"><span class="pre">number_of_heads</span></code> column.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">character_heads</span> <span class="o">=</span> <span class="n">df_hitchhikers</span><span class="p">[</span><span class="s2">&quot;number_of_heads&quot;</span><span class="p">]</span>
<span class="n">character_heads</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>count    5.000000
mean     1.200000
std      0.447214
min      1.000000
25%      1.000000
50%      1.000000
75%      1.000000
max      2.000000
Name: number_of_heads, dtype: float64
</pre></div>
</div>
</div>
</div>
<p>The characters in our DataFrame have at least (min) 1 and at most (max) 2 heads. The mean number of heads is 1.2, but of course this a bit silly, because number of heads is not a continuous variable.</p>
<p><strong>Exercise 3 (English)</strong></p>
<ul class="simple">
<li><p>Access the country column from the <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> DataFrame, and then use the <code class="docutils literal notranslate"><span class="pre">value_counts()</span></code> function to show the number of mosquito samples from each country.</p></li>
<li><p>Access the type column from the <code class="docutils literal notranslate"><span class="pre">df_geneset</span></code> DataFrame, and then use the <code class="docutils literal notranslate"><span class="pre">value_counts()</span></code> function to show the number of genome annotations of different types.</p></li>
<li><p>Access the effect column from the <code class="docutils literal notranslate"><span class="pre">snp_freqs_df</span></code> DataFrame, and then use the <code class="docutils literal notranslate"><span class="pre">value_counts()</span></code> function to show the number of SNPs with different effects.</p></li>
</ul>
<p><strong>Exercice 3 (Franais)</strong></p>
<ul class="simple">
<li><p>Obtenir la colonne country du DataFrame <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> et utiliser la fonction <code class="docutils literal notranslate"><span class="pre">value_counts()</span></code> pour afficher le nombre de moustiques capturs dans chaque pays.</p></li>
<li><p>Obtenir la colonne type du DataFrame <code class="docutils literal notranslate"><span class="pre">df_geneset</span></code> et utiliser la fonction <code class="docutils literal notranslate"><span class="pre">value_counts()</span></code> pour afficher le nombre dannotations du gnome de chaque type.</p></li>
<li><p>Obtenir la colonne effect du DataFrame <code class="docutils literal notranslate"><span class="pre">snp_freqs_df</span></code> et utiliser la fonction <code class="docutils literal notranslate"><span class="pre">value_counts()</span></code> pour afficher le nombre de SNPs pour chaque effet possible.</p></li>
</ul>
</section>
<section id="accessing-multiple-columns">
<h3>Accessing multiple columns<a class="headerlink" href="#accessing-multiple-columns" title="Permalink to this heading">#</a></h3>
<p>To access multiple columns, provide the column names as a list of strings. Lets access the name, taxon and number_of_heads columns from the DataFrame of book characters, and assign the resulting DataFrame to a new variable <code class="docutils literal notranslate"><span class="pre">df_hh</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">selected_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;taxon&quot;</span><span class="p">,</span> <span class="s2">&quot;number_of_heads&quot;</span><span class="p">]</span>
<span class="n">df_hh</span> <span class="o">=</span> <span class="n">df_hitchhikers</span><span class="p">[</span><span class="n">selected_cols</span><span class="p">]</span>
<span class="n">df_hh</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>taxon</th>
      <th>number_of_heads</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Arthur Dent</td>
      <td>human</td>
      <td>1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Tricia McMillan</td>
      <td>human</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Ford Prefect</td>
      <td>betelgeusian</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Zaphod Beeblebrox</td>
      <td>betelgeusian</td>
      <td>2</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Marvin the Paranoid Android</td>
      <td>robot</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>You can also put the list of columns directly inside the square brackets, which would give the same result.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_hh</span> <span class="o">=</span> <span class="n">df_hitchhikers</span><span class="p">[[</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;taxon&quot;</span><span class="p">,</span> <span class="s2">&quot;number_of_heads&quot;</span><span class="p">]]</span>
<span class="n">df_hh</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>taxon</th>
      <th>number_of_heads</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Arthur Dent</td>
      <td>human</td>
      <td>1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Tricia McMillan</td>
      <td>human</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Ford Prefect</td>
      <td>betelgeusian</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Zaphod Beeblebrox</td>
      <td>betelgeusian</td>
      <td>2</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Marvin the Paranoid Android</td>
      <td>robot</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p><strong>Exercise 4 (English)</strong></p>
<ul class="simple">
<li><p>Create a new DataFrame containing only the sample_id, country and year columns from the <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> DataFrame.</p></li>
<li><p>Create a new DataFrame containing only the ID, contig, start and end columns from the <code class="docutils literal notranslate"><span class="pre">df_geneset</span></code> DataFrame.</p></li>
<li><p>Create a new DataFrame containing only the frequency columns from the <code class="docutils literal notranslate"><span class="pre">snp_freqs_df</span></code> DataFrame (hint: the names of these columns all begin with frq_).</p></li>
</ul>
<p><strong>Exercice 4 (Franais)</strong></p>
<ul class="simple">
<li><p>Crer un nouveau DataFrame contenant seulement les colonnes sample_id, country et year du DataFrame <code class="docutils literal notranslate"><span class="pre">df_samples</span></code>.</p></li>
<li><p>Crer un nouveau DataFrame contenant seulement les colonnes ID, contig, start et end du DataFrame <code class="docutils literal notranslate"><span class="pre">df_geneset</span></code>.</p></li>
<li><p>Crer un nouveau DataFrame contenant seulement les colonnes contenant les frquences du DataFrame <code class="docutils literal notranslate"><span class="pre">snp_freqs_df</span></code> (Indice: les noms de ces colonnes commencent tous par frq_).</p></li>
</ul>
</section>
</section>
<section id="accessing-rows">
<h2>Accessing rows<a class="headerlink" href="#accessing-rows" title="Permalink to this heading">#</a></h2>
<p>Now weve seen how to access columns, lets see how to access rows. To access rows by their position within a DataFrame, we can use the <code class="docutils literal notranslate"><span class="pre">iloc</span></code> property.</p>
<section id="accessing-a-single-row">
<h3>Accessing a single row<a class="headerlink" href="#accessing-a-single-row" title="Permalink to this heading">#</a></h3>
<p>To access a single row, we use the <code class="docutils literal notranslate"><span class="pre">iloc</span></code> property, then provide the row number as an integer inside square brackets, remembering that numbering starts from 0 in Python. For example, access the first row of our DataFrame of book characters, and assign the result to a variable <code class="docutils literal notranslate"><span class="pre">first_row</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">first_row</span> <span class="o">=</span> <span class="n">df_hitchhikers</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>When accessing a single row, pandas returns a Series.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">type</span><span class="p">(</span><span class="n">first_row</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>pandas.core.series.Series
</pre></div>
</div>
</div>
</div>
<p>Inspect the data in the row.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">first_row</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>name               Arthur Dent
taxon                    human
gender                    male
number_of_heads              1
intelligence               100
Name: 0, dtype: object
</pre></div>
</div>
</div>
</div>
<p>Access specific data values within the row.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">first_row</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;Arthur Dent&#39;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">first_row</span><span class="p">[</span><span class="s2">&quot;number_of_heads&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>1
</pre></div>
</div>
</div>
</div>
<p><strong>Exercise 5 (English)</strong></p>
<ul class="simple">
<li><p>Access the 81st row in the <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> DataFrame. What country did this mosquito sample come from? (Remember: numbering starts from 0 in Python.)</p></li>
<li><p>Access the 73,287th row in the <code class="docutils literal notranslate"><span class="pre">df_geneset</span></code> DataFrame. What type of annotation is this? What is its Name?</p></li>
</ul>
<p><strong>Exercice 5 (Franais)</strong></p>
<ul class="simple">
<li><p>Obtenir la 81me ligne du DataFrame <code class="docutils literal notranslate"><span class="pre">df_samples</span></code>. De quel pays est issu ce moustique? (Indice: les indices commencent  0 en Python)</p></li>
<li><p>Obtenir la 73 287me ligne du DataFrame <code class="docutils literal notranslate"><span class="pre">df_geneset</span></code>. De quel type dannotation sagit-il? Quel est son nom?</p></li>
</ul>
</section>
<section id="accessing-multiple-rows">
<h3>Accessing multiple rows<a class="headerlink" href="#accessing-multiple-rows" title="Permalink to this heading">#</a></h3>
<p>We can also use <code class="docutils literal notranslate"><span class="pre">iloc</span></code> to access multiple rows,from some start position to some end position. To do this we provide two integers separated by a colon (:) which is called a slice.</p>
<p>As well as remembering that numbering starts from zero, we also have to remember that slices include the start position but do not include the stop position. E.g., to access the first two rows, provide the start and stop positions 0 and 2. This returns a new DataFrame with just the selected rows within the slice.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">first_two_rows</span> <span class="o">=</span> <span class="n">df_hitchhikers</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>
<span class="n">first_two_rows</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>taxon</th>
      <th>gender</th>
      <th>number_of_heads</th>
      <th>intelligence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Arthur Dent</td>
      <td>human</td>
      <td>male</td>
      <td>1</td>
      <td>100</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Tricia McMillan</td>
      <td>human</td>
      <td>female</td>
      <td>1</td>
      <td>180</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p><strong>Exercise 6 (English)</strong></p>
<ul class="simple">
<li><p>Access rows starting from the 6th row up to and including the 10th row in the <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> DataFrame.</p></li>
</ul>
<p><strong>Exercice 6 (Franais)</strong></p>
<ul class="simple">
<li><p>Obtenir toutes les lignes entre la 6me et la 10me (incluses) du DataFrame <code class="docutils literal notranslate"><span class="pre">df_samples</span></code>.</p></li>
</ul>
</section>
</section>
<section id="indexes">
<h2>Indexes<a class="headerlink" href="#indexes" title="Permalink to this heading">#</a></h2>
<p>Every pandas DataFrame has an index, which is a special column (or set of columns) which can be used to access rows quickly and conveniently. In particular, indexes can be used to assign a label to every row, and then use that label to locate and access matching rows.</p>
<p>For example, when working with our DataFrame of book characters, we might want to access rows using the name of the character. To do this, we can set the name column as the index.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_hitchhikers_indexed</span> <span class="o">=</span> <span class="n">df_hitchhikers</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;name&quot;</span><span class="p">)</span>
<span class="n">df_hitchhikers_indexed</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>taxon</th>
      <th>gender</th>
      <th>number_of_heads</th>
      <th>intelligence</th>
    </tr>
    <tr>
      <th>name</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Arthur Dent</th>
      <td>human</td>
      <td>male</td>
      <td>1</td>
      <td>100</td>
    </tr>
    <tr>
      <th>Tricia McMillan</th>
      <td>human</td>
      <td>female</td>
      <td>1</td>
      <td>180</td>
    </tr>
    <tr>
      <th>Ford Prefect</th>
      <td>betelgeusian</td>
      <td>male</td>
      <td>1</td>
      <td>120</td>
    </tr>
    <tr>
      <th>Zaphod Beeblebrox</th>
      <td>betelgeusian</td>
      <td>male</td>
      <td>2</td>
      <td>30</td>
    </tr>
    <tr>
      <th>Marvin the Paranoid Android</th>
      <td>robot</td>
      <td>None</td>
      <td>1</td>
      <td>5000000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>The DataFrame <code class="docutils literal notranslate"><span class="pre">df_hitchhikers_indexed</span></code> has the same data as before, but notice that the name column has moved to the left and is shown in bold. This shows us that the name column is now being used as the DataFrame index.</p>
<p>Now that weve set a useful index, we can use the <code class="docutils literal notranslate"><span class="pre">loc</span></code> property to locate rows with a particular name. E.g., locate the row where the characters name is Tricia McMillan.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">trillian</span> <span class="o">=</span> <span class="n">df_hitchhikers_indexed</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;Tricia McMillan&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>This indexing action locates a single row, and so pandas returns a Series.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">type</span><span class="p">(</span><span class="n">trillian</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>pandas.core.series.Series
</pre></div>
</div>
</div>
</div>
<p>Inspect the row of data we located.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">trillian</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>taxon               human
gender             female
number_of_heads         1
intelligence          180
Name: Tricia McMillan, dtype: object
</pre></div>
</div>
</div>
</div>
<p>Its also useful to know that you can always reset the index back to the default, using the <code class="docutils literal notranslate"><span class="pre">reset_index()</span></code> function. The default is to index the DataFrame with integers.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_hitchhikers_reset</span> <span class="o">=</span> <span class="n">df_hitchhikers_indexed</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">df_hitchhikers_reset</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>taxon</th>
      <th>gender</th>
      <th>number_of_heads</th>
      <th>intelligence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Arthur Dent</td>
      <td>human</td>
      <td>male</td>
      <td>1</td>
      <td>100</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Tricia McMillan</td>
      <td>human</td>
      <td>female</td>
      <td>1</td>
      <td>180</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Ford Prefect</td>
      <td>betelgeusian</td>
      <td>male</td>
      <td>1</td>
      <td>120</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Zaphod Beeblebrox</td>
      <td>betelgeusian</td>
      <td>male</td>
      <td>2</td>
      <td>30</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Marvin the Paranoid Android</td>
      <td>robot</td>
      <td>None</td>
      <td>1</td>
      <td>5000000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p><strong>Exercise 7 (English)</strong></p>
<ul class="simple">
<li><p>Set the sample_id column as the index on the <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> DataFrame, then locate the row with sample identifier BK0055-C. Where was this mosquito collected?</p></li>
<li><p>Set the ID column as the index on the <code class="docutils literal notranslate"><span class="pre">df_geneset</span></code> DataFrame, then locate the row with identifier AGAP000818. What is the name of this gene?</p></li>
</ul>
<p><strong>Exercice 7 (Franais)</strong></p>
<ul class="simple">
<li><p>Faire de la colonne sample_id lindex du DataFrame <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> et ensuite trouver la ligne avec comme identificant dchantillon BK0055-C. Do est issu ce moustique?</p></li>
<li><p>Faire de la colonne ID lindex du DataFrame <code class="docutils literal notranslate"><span class="pre">df_geneset</span></code> et ensuite trouver la ligne avec comme identifiant AGAP000818. Quel est le nom de ce gne?</p></li>
</ul>
</section>
<section id="queries">
<h2>Queries<a class="headerlink" href="#queries" title="Permalink to this heading">#</a></h2>
<p>DataFrames also provide a <code class="docutils literal notranslate"><span class="pre">query()</span></code> function, which can be used to select rows matching some condition or combination of conditions. The <code class="docutils literal notranslate"><span class="pre">query()</span></code> function accepts a string containing a query expression.</p>
<p>Lets query the book characters DataFrame to select only rows where the value of the taxon column is human. To match a value we use the == operator.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_human_hitchhikers</span> <span class="o">=</span> <span class="n">df_hitchhikers</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;taxon == &#39;human&#39;&quot;</span><span class="p">)</span>
<span class="n">df_human_hitchhikers</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>taxon</th>
      <th>gender</th>
      <th>number_of_heads</th>
      <th>intelligence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Arthur Dent</td>
      <td>human</td>
      <td>male</td>
      <td>1</td>
      <td>100</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Tricia McMillan</td>
      <td>human</td>
      <td>female</td>
      <td>1</td>
      <td>180</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Lets do the opposite, selecting only rows where the value of the taxon column is not human, using the != operator.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_non_human_hitchhikers</span> <span class="o">=</span> <span class="n">df_hitchhikers</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;taxon != &#39;human&#39;&quot;</span><span class="p">)</span>
<span class="n">df_non_human_hitchhikers</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>taxon</th>
      <th>gender</th>
      <th>number_of_heads</th>
      <th>intelligence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2</th>
      <td>Ford Prefect</td>
      <td>betelgeusian</td>
      <td>male</td>
      <td>1</td>
      <td>120</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Zaphod Beeblebrox</td>
      <td>betelgeusian</td>
      <td>male</td>
      <td>2</td>
      <td>30</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Marvin the Paranoid Android</td>
      <td>robot</td>
      <td>None</td>
      <td>1</td>
      <td>5000000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Lets query a numerical field using the &gt; greater than operator.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_multi_headed_hitchhikers</span> <span class="o">=</span> <span class="n">df_hitchhikers</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s2">&quot;number_of_heads &gt; 1&quot;</span><span class="p">)</span>
<span class="n">df_multi_headed_hitchhikers</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>taxon</th>
      <th>gender</th>
      <th>number_of_heads</th>
      <th>intelligence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>3</th>
      <td>Zaphod Beeblebrox</td>
      <td>betelgeusian</td>
      <td>male</td>
      <td>2</td>
      <td>30</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>A query expression can combine conditions on multiple columns using and and or. E.g., find rows where the value of the taxon column is human and the value of the number_of_heads column is 1.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_non_human_one_headed_hitchhikers</span> <span class="o">=</span> <span class="n">df_hitchhikers</span><span class="o">.</span><span class="n">query</span><span class="p">(</span>
    <span class="s2">&quot;taxon != &#39;human&#39; and number_of_heads == 1&quot;</span>
<span class="p">)</span>
<span class="n">df_non_human_one_headed_hitchhikers</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>taxon</th>
      <th>gender</th>
      <th>number_of_heads</th>
      <th>intelligence</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2</th>
      <td>Ford Prefect</td>
      <td>betelgeusian</td>
      <td>male</td>
      <td>1</td>
      <td>120</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Marvin the Paranoid Android</td>
      <td>robot</td>
      <td>None</td>
      <td>1</td>
      <td>5000000</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p><strong>Exercise 8 (English)</strong></p>
<ul class="simple">
<li><p>Query the <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> DataFrame to select mosquitoes collected in Burkina Faso.</p></li>
<li><p>Query the <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> DataFrame to select mosquitoes collected in Burkina Faso in 2014.</p></li>
<li><p>Query the <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> DataFrame to select Anopheles coluzzii mosquitoes collected in Burkina Faso in 2014 (hint: <code class="docutils literal notranslate"><span class="pre">taxon</span> <span class="pre">==</span> <span class="pre">'coluzzii'</span></code>).</p></li>
<li><p>Query the <code class="docutils literal notranslate"><span class="pre">snp_freqs_df</span></code> DataFrame to select SNPs which change the protein sequence (hint: <code class="docutils literal notranslate"><span class="pre">effect</span> <span class="pre">==</span> <span class="pre">'NON_SYNONYMOUS_CODING'</span></code>).</p></li>
<li><p>Query the <code class="docutils literal notranslate"><span class="pre">snp_freqs_df</span></code> DataFrame to select SNPs which change the protein sequence and which occur above 5% frequency in at least one cohort (hint: <code class="docutils literal notranslate"><span class="pre">max_af</span> <span class="pre">&gt;</span> <span class="pre">0.05</span></code>).</p></li>
</ul>
<p><strong>Exercice 8 (Franais)</strong></p>
<ul class="simple">
<li><p>Slectionner parmi les moustiques du DataFrame <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> ceux qui ont t capturs au Burkina Faso.</p></li>
<li><p>Slectionner parmi les moustiques du DataFrame <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> ceux qui ont t capturs au Burkina Faso en 2014.</p></li>
<li><p>Slectionner parmi les moustiques du DataFrame <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> les Anophles coluzzii qui ont t capturs au Burkina Faso en 2014. (Indice: <code class="docutils literal notranslate"><span class="pre">taxon</span> <span class="pre">==</span> <span class="pre">'coluzzii'</span></code>).</p></li>
<li><p>Slectionner parmi les SNPs du DataFrame <code class="docutils literal notranslate"><span class="pre">snp_freqs_df</span></code> ceux qui modifient la squence dune protine (Indice: <code class="docutils literal notranslate"><span class="pre">effect</span> <span class="pre">==</span> <span class="pre">'NON_SYNONYMOUS_CODING'</span></code>).</p></li>
<li><p>Slectionner parmi les SNPs du DataFrame <code class="docutils literal notranslate"><span class="pre">snp_freqs_df</span></code> ceux qui modifient la squence dune protine et apparaissent avec une frquence suprieure  5% dans au moins une cohorte (Indice: <code class="docutils literal notranslate"><span class="pre">max_af</span> <span class="pre">&gt;</span> <span class="pre">0.05</span></code>).</p></li>
</ul>
</section>
<section id="input-output">
<h2>Input/output<a class="headerlink" href="#input-output" title="Permalink to this heading">#</a></h2>
<p>We saw at the beginning of this module how to read in data from a file. Lets look finally at how to write data out to a file.</p>
<p>To write a DataFrame out to a CSV file, use the <a class="reference external" href="https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html">to_csv()</a> function. E.g., lets write out the book characters DataFrame to a CSV file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_hitchhikers</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;hitchhikers.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>You can download this file to your local computer by opening the file browser tab to the left in colab.</p>
<p>There are several other file formats you can write to. E.g., you could write to an Excel file. Here, we use the</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_hitchhikers</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="s2">&quot;hitchhikers.xlsx&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Any of the DataFrames we created in this module could be written out in the same way. E.g., we could write a DataFrame that we obtained by querying the book characters for only humans.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">df_human_hitchhikers</span><span class="o">.</span><span class="n">to_excel</span><span class="p">(</span><span class="s2">&quot;hh_human.xlsx&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Exercise 9 (English)</strong></p>
<ul class="simple">
<li><p>Write the <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> DataFrame out to an Excel file.</p></li>
<li><p>Query the <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> DataFrame to select only mosquitoes from Burkina Faso, then write the resulting DataFrame out to a CSV file.</p></li>
</ul>
<p><strong>Exercice 9 (Franais)</strong></p>
<ul class="simple">
<li><p>Enregistrer le DataFrame <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> sous forme de fichier Excel.</p></li>
<li><p>Slectionner parmi les moustiques du DataFrame <code class="docutils literal notranslate"><span class="pre">df_samples</span></code> ceux provenant du Burkina Faso et enregistrer le DataFrame ainsi cr sous forme de fichier CSV.</p></li>
</ul>
</section>
<section id="topics-not-covered-here">
<h2>Topics not covered here<a class="headerlink" href="#topics-not-covered-here" title="Permalink to this heading">#</a></h2>
<p>Please note that this module has only covered a few of the many features that pandas supports. Here are a few of the features that we have not covered, but which you might like to learn more about through other online tutorials.</p>
<ul class="simple">
<li><p>Grouping and aggregation</p></li>
<li><p>Pivot tables</p></li>
<li><p>Joining and merging</p></li>
<li><p>Concatenating</p></li>
<li><p>Advanced indexing (multi-level indexes; Boolean indexing; )</p></li>
<li><p>Modifying data</p></li>
<li><p>Apply</p></li>
<li><p>Missing data</p></li>
<li><p>Time series</p></li>
<li><p>Plotting</p></li>
<li><p>Text</p></li>
<li><p></p></li>
</ul>
</section>
<section id="further-reading">
<h2>Further reading<a class="headerlink" href="#further-reading" title="Permalink to this heading">#</a></h2>
<p>Here are a few links to online resources providing further information about pandas for new users.</p>
<ul class="simple">
<li><p><a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html">pandas getting started</a></p></li>
<li><p><a class="reference external" href="https://pandas.pydata.org/docs/user_guide/index.html">pandas user guide</a> (in particular see <a class="reference external" href="https://pandas.pydata.org/docs/user_guide/10min.html">10 minutes to pandas</a>)</p></li>
<li><p><a class="reference external" href="https://pandas.pydata.org/pandas-docs/stable/getting_started/tutorials.html">pandas community tutorials</a></p></li>
</ul>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "conda-env-developer-developer-training-nb-maintenance-mgen-7.10.0-py"
        },
        kernelOptions: {
            name: "conda-env-developer-developer-training-nb-maintenance-mgen-7.10.0-py",
            path: "./workshop-2"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'conda-env-developer-developer-training-nb-maintenance-mgen-7.10.0-py'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="about.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Workshop 2 - Metabolic insecticide resistance and copy number variation</p>
      </div>
    </a>
    <a class="right-next"
       href="module-2-metabolic-ir.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Module 2 - The genetic basis of metabolic resistance to insecticides in <em>Anopheles</em> mosquitoes</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#learning-objectives">Learning objectives</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#lecture">Lecture</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#english">English</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#francais">Franais</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup">Setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pandas-dataframes">pandas DataFrames</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-sample-metadata">Example: sample metadata</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-genome-annotations">Example: genome annotations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-snp-allele-frequencies">Example: SNP allele frequencies</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#aside-variable-naming-styles">Aside: variable naming styles</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-reading-from-a-csv-file">Example: reading from a CSV file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-creating-a-dataframe-from-scratch">Example: creating a DataFrame from scratch</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#structure-of-a-dataframe">Structure of a DataFrame</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#counting-rows-and-columns">Counting rows and columns</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-columns">Accessing columns</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-column-names">Accessing column names</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-a-single-column">Accessing a single column</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#investigating-data-in-a-column">Investigating data in a column</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-multiple-columns">Accessing multiple columns</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-rows">Accessing rows</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-a-single-row">Accessing a single row</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#accessing-multiple-rows">Accessing multiple rows</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#indexes">Indexes</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#queries">Queries</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input-output">Input/output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#topics-not-covered-here">Topics not covered here</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#further-reading">Further reading</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By MalariaGEN and PAMCA
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
       Copyright 2022.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
<div class="extra_footer">
  <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons Licence" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a> This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>.

</div>
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>