{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "18jg3kAn8HTx"
   },
   "source": [
    "![banner](https://anopheles-genomic-surveillance.github.io/_images/banner.jpg)\n",
    "\n",
    "***Training course in data analysis for genomic surveillance of African malaria vectors***\n",
    "\n",
    "**[Workshop 7](about) - Gene flow and the spread of insecticide resistance**\n",
    "\n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "d-c9I1DJ8NP1"
   },
   "source": [
    "# Module 4 - Investigating gene flow at loci of interest using haplotype networks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Rg5fvByN8Q63"
   },
   "source": [
    "__Theme: Analysis__\n",
    "\n",
    "\n",
    "In module 2 of workshop 7, we learnt about __selective sweeps__ being shared between populations through __adaptive gene flow__, and in and module 3 we used hierarchical haplotype clustering to investigate and visualise haplotypes spreading between countries and species, allowing us to track gene flow at insecticide resistance loci. We also have another methodology in our armamentarium which lets us visualise the relationship between haplotypes in a different way, and that is __haplotype networks__. In this module we will learn how to create and interpret haplotype networks to analyse gene flow at loci associated with insecticide resistance.\n",
    "\n",
    "**Please note: This module contains interactive plots using Plotly Dash and Cytoscape.js. These plots will not be visible on the course website, but can be generated by running the notebook for yourself. To run this notebook, click the rocket icon at the top of the page then select \"Colab\".**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jp5HIXQH8W8e"
   },
   "source": [
    "## Learning objectives\n",
    "- Understand the theory behind haplotype networks.\n",
    "- Use the malariagen_data python package to build haplotype networks for genomic regions with evidence of adaptive gene flow.\n",
    "- Learn how to interpret haplotype networks to find evidence for gene flow events between countries or species."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lecture"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### English"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/C8LRBAq9Hpw\" title=\"YouTube video player\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%html\n",
    "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/C8LRBAq9Hpw\" title=\"YouTube video player\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Français"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": [
     "remove-input"
    ]
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/PJe1wSSo19Y\" title=\"YouTube video player\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%html\n",
    "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/PJe1wSSo19Y\" title=\"YouTube video player\" frameborder=\"0\" allow=\"accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Please note that the code in the cells below might differ from that shown in the video. This can happen because Python packages and their dependencies change due to updates, necessitating tweaks to the code."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "6_X-rLBv8XGN"
   },
   "source": [
    "## Setup\n",
    "\n",
    "First, let’s begin by installing and importing some Python packages, and configuring access to _Anopheles_ genomic data from the MalariaGEN Ag3.0 data resource."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "id": "ytSj2yhY6lp4"
   },
   "outputs": [],
   "source": [
    "!pip install -q --no-warn-conflicts malariagen_data dash==2.9.3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "uh4SQVAP6q0u"
   },
   "outputs": [],
   "source": [
    "import malariagen_data\n",
    "import dash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 250
    },
    "id": "R97e6Z7f6s1y",
    "outputId": "b9dbf143-cecb-42fc-d82f-f0f36a3fd5eb"
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.2.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.2.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.2.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.2.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.2.0.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "          for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.2.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.2.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.2.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.2.0.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.2.0.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "            <table class=\"malariagen-ag3\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\" colspan=\"2\">MalariaGEN Ag3 API client</th>\n",
       "                    </tr>\n",
       "                    <tr><td colspan=\"2\" style=\"text-align: left\">\n",
       "                        Please note that data are subject to terms of use,\n",
       "                        for more information see <a href=\"https://www.malariagen.net/data\">\n",
       "                        the MalariaGEN website</a> or contact data@malariagen.net.\n",
       "                        See also the <a href=\"https://malariagen.github.io/vector-data/ag3/api.html\">Ag3 API docs</a>.\n",
       "                    </td></tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Storage URL\n",
       "                        </th>\n",
       "                        <td>gs://vo_agam_release/</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Data releases available\n",
       "                        </th>\n",
       "                        <td>3.0</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Results cache\n",
       "                        </th>\n",
       "                        <td>None</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Cohorts analysis\n",
       "                        </th>\n",
       "                        <td>20230516</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            AIM analysis\n",
       "                        </th>\n",
       "                        <td>20220528</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Site filters analysis\n",
       "                        </th>\n",
       "                        <td>dt_20200416</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Software version\n",
       "                        </th>\n",
       "                        <td>malariagen_data 7.12.0</td>\n",
       "                    </tr>\n",
       "                    <tr>\n",
       "                        <th style=\"text-align: left\">\n",
       "                            Client location\n",
       "                        </th>\n",
       "                        <td>unknown</td>\n",
       "                    </tr>\n",
       "                </tbody>\n",
       "            </table>\n",
       "        "
      ],
      "text/plain": [
       "<MalariaGEN Ag3 API client>\n",
       "Storage URL             : gs://vo_agam_release/\n",
       "Data releases available : 3.0\n",
       "Results cache           : None\n",
       "Cohorts analysis        : 20230516\n",
       "AIM analysis            : 20220528\n",
       "Site filters analysis   : dt_20200416\n",
       "Software version        : malariagen_data 7.12.0\n",
       "Client location         : unknown\n",
       "---\n",
       "Please note that data are subject to terms of use,\n",
       "for more information see https://www.malariagen.net/data\n",
       "or contact data@malariagen.net. For API documentation see \n",
       "https://malariagen.github.io/vector-data/ag3/api.html"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ag3 = malariagen_data.Ag3()\n",
    "ag3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "FNwlDAtj8hFA"
   },
   "source": [
    "## Haplotype networks\n",
    "\n",
    "We have an experimental function for building haplotype networks in the `malariagen_data` python package (we're still ironing out some bugs!). Let's have a look at the doc string for the `ag3.plot_haplotype_network()` function by using \"?\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "id": "seLA0-Uy8itr"
   },
   "outputs": [],
   "source": [
    "ag3.plot_haplotype_network?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "94ZEKRRK8l_y"
   },
   "source": [
    "The required parameters, `region` and `analysis` are familiar to us by now, but there are a some new optional parameters here that might be useful to understand.\n",
    "\n",
    "- `max_dist=` This is the maximum distance in _n_ SNPs where the algorithm will join haplotype nodes together as one network, after which it will split the nodes in to separate networks. The default here is 2 SNPs. We will see an example of how this works shortly.\n",
    "- `color=` Identifies a column in the sample metadata which determines the colour of pie chart segments within nodes. _e.g._ we could use \"country\" or \"taxon\" columns.\n",
    "\n",
    "It is also worth knowing that if you get __403 errors__ instead of figures, try reloading your browser tab or restart your browser."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "wtnPDX1DyYA2"
   },
   "source": [
    "Let's build a haplotype network around _Kdr_ locus in the _Vgsc_ gene for all the _An. coluzzii_ samples in Ag3.0, and look at the output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 735,
     "referenced_widgets": [
      "48d3d0a75d21416e9ffc1a8945f6576c",
      "4365a7b25cdc4e2fb00d013d7f0d8382",
      "326ce47f022042afac80ad5b42d84def",
      "5fd44d5089fe484aa2037e92a3413170",
      "ddd49e71ff1644f6a7f74f6c2a0eee37",
      "b83b1b6e58114cdba8cbece1ec12e706",
      "4b14f10cea454dd2bbcbc2ae85bab6ff",
      "3a8dd67a5d804ca3815e7500128dd2c9",
      "29825bc2a25b4b559a5b6c33ee06a767",
      "9a6079af3170459fbf5f94cc4d3327f6",
      "682a3755b1ff4b64aeacd68f0b318c32",
      "b5dca63364044c49babcafe6259cb2d7",
      "db7b2f501e064d90ad6f7055fe9e6459",
      "54b580d5b0da461dac5666b24728f765",
      "c855dcdb372b4c1eaa3733e527257d15",
      "4d96e6a185494a09bf9cbf43d0058f78",
      "52cf5f2426da499abde1e1cc87b64ba8",
      "c9a61a460d6542769aee3d4ad53bce77",
      "a196afa55af04a72812bd6629d022d0d",
      "cba8a81f9ef14172a27dc8229b1a0a2c",
      "c26f6c30a267415ca4265d5352ef8878",
      "c1496a8c06bd4300bbc9d7b3dd9adca8"
     ]
    },
    "id": "tJ2E5dVRbF2s",
    "outputId": "9d03cfc2-35f3-4c47-ba11-0d5d187aee83"
   },
   "outputs": [],
   "source": [
    "ag3.plot_haplotype_network(\n",
    "    region='2L:2,400,000-2,440,000',\n",
    "    analysis='gamb_colu',\n",
    "    sample_query= 'taxon == \"coluzzii\"',\n",
    "    color=\"country\",\n",
    "    max_dist=2,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "G1cm_vmIz89L"
   },
   "source": [
    "### Technical interpretation\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "_n_R-a0DyCV5"
   },
   "source": [
    "The output figure represents identical haplotypes as **nodes** (circles), the larger the node, the greater the number of identical haplotypes. In this example, the `color` parameter has been set to the \"country\" column of the haplotype metadata, so the node colour proportions represent the countries where that haplotype was present in our data set. We often also set the `color` parameter to \"taxon\", in which case, node colour proportion represents the species which carried that haplotype. If we click or tap a node, information about its composition is provided at the bottom of the figure. This information relates the the `color` parameter, _i.e._ if `color=taxon`, the information will be about species composition. Nodes can also be moved to aid visualisation (`shift + click` to select multiple), and the whole figure can be zoomed in and out using the mouse wheel.\n",
    "\n",
    "The **edges** (lines) that join nodes together represent the genetic distance between the nodes in SNPs. Each edge is a single SNP difference.\n",
    "\n",
    "\n",
    "In the _Vgsc_ example, with the `max_dist` (maximum distance) parameter set to two, we se each node in these networks is separated by a maximum genetic distance of 2 SNPs, shown as two edges with a small grey node in the middle. If we click on a small grey node we see \"No. haplotypes: 0\". The evolutionary relationship between these haplotype nodes either side of the small node has been inferred as we don't actually have the intermediate haplotype (with only one SNP difference) in our data set. Let's increase the `max_dist` parameter and see what happens."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 703,
     "referenced_widgets": [
      "bad68646be3a465aaa61a87c20efea54",
      "1aa24e6a6f604247972aa908ccf500ab",
      "9318271dcf7744408bf6572f2f14828d",
      "78b148ec94114cff8d2691c233b8be6e",
      "b4844f3093dc428087b1eb12416d4ff9",
      "ec699fb70d664df4a77d9e4b6eb622d9",
      "12680adb135b4f6db9212b31169345a3",
      "e16b46cd1da243afad758d1d43e198f6",
      "cbb202011e8940e69cc098dd290590ce",
      "f10a9ddc583240dda711a5e91c117f4c",
      "c32cc806db2a414faf7758ab3724e9aa"
     ]
    },
    "id": "gkxnO7LaRf1V",
    "outputId": "d4268a45-b116-4e26-960a-1d97054e56b1"
   },
   "outputs": [],
   "source": [
    "ag3.plot_haplotype_network(\n",
    "    region='2L:2,400,000-2,440,000',\n",
    "    analysis='gamb_colu',\n",
    "    sample_query= 'taxon == \"coluzzii\"',\n",
    "    color=\"country\",\n",
    "    max_dist=4,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "KaAylb8jRpEL"
   },
   "source": [
    "If we increase the `max_dist` parameter to four SNPs, we see haplotypes linked together by up to three \"intermediate\" node, the network has found connections between more distantly related haplotypes. We generally use `max_dist=2`, mainly because we're interested here only in groups of very closely related haplotypes, which are therefore very likely to share a recent common origin.\n",
    "  \n",
    "To aid interpretation, **only networks containing more than one haplotype are represented in the output figure**. This means that if a particular haplotype is found only once in our data set (_i.e._ a singleton haplotype), and that haplotype is more than the `max_dist` number of SNPs different from any other haplotype, it not be shown."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "roBro0yG0_Qa"
   },
   "source": [
    "### Evolutionary interpretation\n",
    "\n",
    "Let's return to the _Vgsc_ network example with `max_dist=2` and try to interpret the evolutionary inference with respect to insecticide resistance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 703,
     "referenced_widgets": [
      "d1514031c1fc4be7a871514a703fad8c",
      "7172c22b8a9f40a2916398788d06a549",
      "75c51d58416d445ca65c05ea72657fb1",
      "edc848b4d4e44425aab1a465f19c5337",
      "f1207ec651874ac0a23300801408a582",
      "50bba9a71c384ae88da83eac6092d244",
      "14080bf572dc48918e6c0384751ad73f",
      "196e1acfeb9f4c7f992bb6b80827a61a",
      "90106d3f3bba4891858dade1d4d4fa36",
      "637f0e94b3064a6bba3320cc107faffb",
      "bf383caab42648008e147e86862982a8"
     ]
    },
    "id": "untae-dA4ZEY",
    "outputId": "f63ccb06-dfd2-479d-a7cf-451774c4d6ec"
   },
   "outputs": [],
   "source": [
    "ag3.plot_haplotype_network(\n",
    "    region='2L:2,400,000-2,440,000',\n",
    "    analysis='gamb_colu',\n",
    "    sample_query= 'taxon == \"coluzzii\"',\n",
    "    color=\"country\",\n",
    "    max_dist=2,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "h97JEXzL4oVP"
   },
   "source": [
    "\n",
    "\n",
    "- When a fitness increasing mutation occurs, haplotype it resides on comes under selection and it increases in frequency. So, here we're interested especially in the large nodes, because these are very likely to be __high frequency haplotypes experiencing positive selection__.\n",
    "\n",
    "- During a selective sweep, some additional mutations will occur on the original haplotype. These could be neutral or they could be secondary functional mutations (_i.e._ further increasing fitness), but key point is that these additional mutations occur, and give us a network of closely related haplotypes. We can infer that each network of connected haplotypes have a recent common origin.\n",
    "\n",
    "- If a network of connected haplotypes contains haplotypes from two or more countries, this is evidence for adaptive gene flow. Particularly a large node containing haplotypes from multiple countries, this is very strong evidence for adaptive gene flow because exactly the same haplotype is found across countries.\n",
    "\n",
    "- Likewise, if a network of connected haplotypes contains haplotypes from two or more species, this is also evidence for a type of adaptive gene flow called __adaptive introgression__. Again, a large node containing haplotypes from multiple species, is very strong evidence for adaptive gene flow.\n",
    "\n",
    "\n",
    "The most striking feature of this _Vgsc_ network is the very large and complex network at the top the figure. The larger node colours show that there are several high frequency haplotypes, closely related to each other, shared across countries. This is a classic signal of adaptive gene flow and in this case, we know from previous research, the haplotypes all carry the L995F _kdr_ mutations that confers resistance to pyrethroids and DDT. Furthermore, the multiple large nodes suggest that secondary mutations have recently occurred on an already selected haplotype, then these new haplotypes have increased in frequency, likely also under selection.\n",
    "\n",
    "We can also see some smaller complex networks, which suggest potential selective sweeps, two of which show evidence of adaptive gene flow between countries (sharing of identical haplotypes).\n",
    "\n",
    "The smallest networks are fairly uninformative, and could be considered noise.\n",
    "\n",
    "Now we have an idea how to run and interpret haplotype networks, let's look at some insecticide resistance loci we looked at in module 3."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "A0frHcX2CAao"
   },
   "source": [
    "## Exercise 1\n",
    "\n",
    "### English\n",
    "\n",
    "- Repeat the _Vgsc_ network analysis, but this time investigate adaptive gene flow (introgression) between _An. gambiae_ and _coluzzii_. _HINT_: `color=?????`.\n",
    "- Is there evidence of adaptive gene flow between species?\n",
    "\n",
    "### Français\n",
    "\n",
    "- Répéter l'analyse des réseaux pour _Vgsc_ mais cette fois étudier le flux génétique adaptatif (introgression) entre _An. gambiae_ et _coluzzii_. _INDICE_: `color=?????`.\n",
    "- Y a-t-il des indications de flux génétique adaptatif entre espèces?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "S5bR_U239ACp"
   },
   "source": [
    "## Example: _Ace1_\n",
    "\n",
    "We've heard about the _Ace1_ genes involvement in organophosphate insecticide resistance, and in the last module, we've also seen the results of visualising the relationships between haplotypes at this locus using __hierarchical clustering__. Let's first remind ourselved what that looked like for _Ace1_ with _An. gambiae_ haplotypes from from Ghana and Burkina Faso..\n",
    "\n",
    "### _Ace1_ hierarchical clustering re-cap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 569,
     "referenced_widgets": [
      "8841f6eae064408eab59668729c4b756",
      "a04fd141f40b4e4481722e2701b21628",
      "82877e72ed144b8f80568536717219dd",
      "0b348a3cdb784529b88e18e1fdb56d2f",
      "c4f183a81a2f4c038bedd894f2b4cda8",
      "e525d3c6f192466bb4cb6d978738c07c",
      "37cc3979ccf74d32af230fe0e499104b",
      "4ec8613701be44cca1059189466674f0",
      "9df49397bc974f089749b505235cfbac",
      "0803db7a67964b20ac7a885612b48fa4",
      "ffdd1d3a3acd477aab73128f05396075"
     ]
    },
    "id": "zvNlxaIpT3Wr",
    "outputId": "dd99ebe0-ec9b-4d87-9204-d4ee223ae1a5"
   },
   "outputs": [],
   "source": [
    "ag3.plot_haplotype_clustering(\n",
    "    region='2R:3,470,000-3,490,000',\n",
    "    analysis='gamb_colu',\n",
    "    sample_query= 'country in [\"Ghana\", \"Burkina Faso\"] & taxon == \"gambiae\"',\n",
    "    color='country')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "qJkyABhJS_Tp"
   },
   "source": [
    "Towards the right hand side of the dendrogram, we can see a region where there are haplotypes closely related to each other. By zooming in we can find that this distance is <5 SNPs between haplotypes, rather than >50 SNPs distance we see elsewhere in the figure. This pattern is suggestive of a selective sweep, and as it includes multiple countries, is also suggestive of __adaptive gene flow__. \n",
    "\n",
    "Adaptive gene flow is where a fitness improving mutation increases in frequency and spreads across space, between populations and/or between different species. Here we are detecting the increase and spread of the haplotype which carries this locus under positive selection. Given the importance of this gene to insecticide action and the high evolutionary pressure that these chemicals put on mosquitoes, this signal may well be driven by a resistance phenotype associated locus. \n",
    "\n",
    "Let's investigate the same region but using haplotype networks. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "CAEeDphSFwa_"
   },
   "source": [
    "### _Ace1_ haplotype networks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 703,
     "referenced_widgets": [
      "d909adffe1f64bcfb286b285d60103b6",
      "bb0e6e043e274c929793e5b658eb06a0",
      "9eb75ed8464b423785eb1732a41301c6",
      "f17cb7e1937d427e95f68e002420c117",
      "2a75bff9a03f46cda2ec4936ebc60082",
      "0bac59e4e9c1435c87661652de854a53",
      "e91454f2b8fa4bcfac205e11f514f832",
      "b32268c64c3642cfb8d8af6970fa5916",
      "83788f3f6f664711bcdf076a8fc3c981",
      "101e880a18ff4155b919031429bca245",
      "fc67766129774380aa27b0431f0925f6"
     ]
    },
    "id": "aORNA8_A6wIk",
    "outputId": "ebad998a-6d92-4481-dc6b-580bf8454f18"
   },
   "outputs": [],
   "source": [
    "ag3.plot_haplotype_network(\n",
    "    region='2R:3,470,000-3,490,000',\n",
    "    analysis='gamb_colu',\n",
    "    sample_query= 'country in [\"Ghana\", \"Burkina Faso\"] & taxon == \"gambiae\"',\n",
    "    color=\"country\",\n",
    "    max_dist=2,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "dJgDqaUJfvTa"
   },
   "source": [
    "Looking back at the hierarchical clustering analysis of _Ace1_, the vast majority of the dendrogram leaves (haplotypes) have long branches (genetic distances) to their nearest neighbours. All of those singleton haplotypes, more that 2 SNPs distance from any other haplotypes, are filtered out in the network analysis as we are only interested in high frequency haplotypes, and networks of closely related haplotypes, because they are likely to be haplotypes under recent positive selection. What remains are the haplotypes which make us the networks we can see above, a much simplified picture of the genetic relationships.\n",
    "\n",
    "The largest network corresponds to the selective sweep signal we saw in the clustering analysis. Clicking on the largest node, we see it contains 15 identical haplotypes, 11 from Burkina Faso and four from Ghana. As expected, this corresponds perfectly with what we see if we zoom in to the far right side of the clustering figure and roll over the leaves with our mouse, further evidence this haplotype has spread, under selection between countries. Several larger nodes, closely related to each other gives us evidence that the selection is recent (as we might expect if driven by malaria control insecticidal pressures)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "PyDjeE4lIAUv"
   },
   "source": [
    "## Exercise 2\n",
    "\n",
    "### English\n",
    "\n",
    "For _Ace1_ add _coluzzii_ samples as well, colour the nodes by taxon and take another look at these networks. _HINT_ `sample_query= 'country in [\"Ghana\", \"Burkina Faso\"] & taxon in [\"gambiae\", \"coluzzii\"]'`\n",
    "- is there evidence of adaptive introgression of haplotypes between species at the _Ace1_ locus?\n",
    "\n",
    "### Français\n",
    "\n",
    "Pour _Ace1_ ajouter les _coluzzii_, colorer les noeuds par taxon et regarder à nouveau les réseaux. _INDICE_ `sample_query= 'country in [\"Ghana\", \"Burkina Faso\"] & taxon in [\"gambiae\", \"coluzzii\"]'`\n",
    "- Y a-t-il des indications d'introgression adaptive entre espèces au locus _Ace1_ ?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "4QJsw3D-drKE"
   },
   "source": [
    "## Example: _Gste2_\n",
    "\n",
    "We have seen that Gste2 gene is associated with insecticide resistance, and we saw evidence for adaptive gene flow at this locus using hierarchical clustering. Lets now perform a haplotype networking analysis at this gene and compare the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 569,
     "referenced_widgets": [
      "162f1e4844fa4569a02d16255d60f9e1",
      "d964fb7594624556873bddad8bd13367",
      "addefa88982444a8b6a5ef608fd55796",
      "37afa36c5b524a9d9fdadf2cccc34d9b",
      "493e1437e4e04c25817c3877607e42ba",
      "9f09c3ca873247b4acdc890040f7b427",
      "16e212d2b88e4f9db60427393807972a",
      "940e404dad3c4202b5f744d4fadf10fa",
      "870f866a51ce4b6a943884db318fa668",
      "505c1d1ed22443ea8f6e4a574c331728",
      "e1faed0d9a3a4f50b9070f4a4b9de320"
     ]
    },
    "id": "9Xm6AtBWSI77",
    "outputId": "ac3afc13-c354-44dc-a503-cb552926b135"
   },
   "outputs": [],
   "source": [
    "ag3.plot_haplotype_clustering(\n",
    "    region='3R:28,590,000-28,600,000',\n",
    "    analysis='gamb_colu',\n",
    "    sample_query= 'country in [\"Ghana\", \"Burkina Faso\", \"Cote d\\'Ivoire\"]',\n",
    "    color='taxon',\n",
    "    symbol='country'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 703,
     "referenced_widgets": [
      "d674d26d0f224869a4c02f9a464675b9",
      "647fb49da93a41b29adc6c440da667b3",
      "94a865c39c464ffaa8c0a9859d0b354d",
      "9c9ea6330a684fbfa8e3293cdd47d8e1",
      "dfaddf2b06dc47babcf96763e890604f",
      "b1509ae7a95c4e7e8ed11076b89b13b0",
      "b12da32cd5dc468ea3f73f811493ebb8",
      "73450e7dbff346f791f528828c2e300c",
      "6f30a4cdd40f450aa75a8739c2545fec",
      "9cebe01ba7714663ba7e345dcbc8d94b",
      "96a89dc19ea24de18e064a46957bde3d"
     ]
    },
    "id": "3eJb285GqMrG",
    "outputId": "0d902580-f49d-406c-b81b-01ef5a333aa4"
   },
   "outputs": [],
   "source": [
    "ag3.plot_haplotype_network(\n",
    "    region='3R:28,590,000-28,600,000',\n",
    "    analysis='gamb_colu',\n",
    "    sample_query= 'country in [\"Ghana\", \"Burkina Faso\", \"Cote d\\'Ivoire\"]',\n",
    "    color=\"country\",\n",
    "    max_dist=2,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 703,
     "referenced_widgets": [
      "470dacaa5aee42559fb8b99ecb017d55",
      "721685a3abc748d5a4efb871e51d8582",
      "22ea8062d274450687941082a4fce64d",
      "e8f92d260bf04e94add2c91973c7186b",
      "dabae4399dad4aef88a9a9cac5264a87",
      "f93a548be11d4149a59846fe60d7588e",
      "7d87936965cb4f749f1659fac7b2b828",
      "6955152fe9a84f5a9f23716a55678cd5",
      "9bb5aa113a714cf1a87f8465699322f8",
      "33f8546eaf6141c6a4a4bbe2e0babd50",
      "f73b66c1e27f4e169290effe70ddc25e"
     ]
    },
    "id": "ZvhaR7gTTWRM",
    "outputId": "9830a6eb-f840-4f07-e740-8edc819f5f30"
   },
   "outputs": [],
   "source": [
    "ag3.plot_haplotype_network(\n",
    "    region='3R:28,590,000-28,600,000',\n",
    "    analysis='gamb_colu',\n",
    "    sample_query= 'country in [\"Ghana\", \"Burkina Faso\", \"Cote d\\'Ivoire\"]',\n",
    "    color=\"taxon\",\n",
    "    max_dist=2,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "8F6VbmkcTr5p"
   },
   "source": [
    "The haplotype network figures make it easy to spot the selective sweeps and gene flow between Burkina Faso and Ghana (but not Cote d'Ivoire) and the adaptive introgression of haplotypes between _gambiae_ and _coluzzii_ at this locus.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "DVn3AYeXUD1B"
   },
   "source": [
    "## Exercise 3\n",
    "\n",
    "### English\n",
    "\n",
    "The cytochrome P450 gene _Cyp9k1_ is associated with metabolic pyrethroid resistance. Run a haplotype clustering analysis on samples from Ghana and Cote d'Ivoire on this region. _HINT_:  region='X:15,240,000-15,250,000'\n",
    "\n",
    "- Is there evidence of selection?\n",
    "- Is there evidence of adaptive gene flow?\n",
    "- If so, between countries and/or species?\n",
    "\n",
    "### Français\n",
    "\n",
    "Le cytochrome P450 gene _Cyp9k1_ est associé à la résistance métabolique aux pyrethrinoïdes. Exécuter une analyse par regroupement d'haplotypes pour les échantillons du Ghana et de la Côte d'Ivoire dans cette région. _INDICE_: region='X:15,240,000-15,250,000'\n",
    "\n",
    "- Y a-t-il des indications de sélection?\n",
    "- Y a-t-il des indications de flux génétique adaptatif?\n",
    "- Si oui, entre pays et/ou espèces?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "WK7zehBOEEuR"
   },
   "source": [
    "## Well done!\n",
    "\n",
    "In this module we have:  \n",
    "- Learnt about the theory behind haplotype networks.\n",
    "- Used the malariagen_data python package to build haplotype networks for genomic regions with evidence of adaptive gene flow.\n",
    "- Learnt how to interpret haplotype networks to discern movement of insecticide resistance loci between countries and species."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ndOMHqNvEmcy"
   },
   "source": [
    "## Exercises\n",
    "\n",
    "### English\n",
    "\n",
    "Open this notebook in Google Colab and run it for yourself from top to bottom. As you run through the notebook, cell by cell, think about what each cell is doing, and try the practical exercises along the way.\n",
    "\n",
    "Have go at the practical exercises, but please don’t worry if you don’t have time to do them all during the practical session, and please ask the teaching assistants for help if you are stuck.\n",
    "\n",
    "Hint: To open the notebook in Google Colab, click the rocket icon at the top of the page, then select “Colab” from the drop-down menu.\n",
    "\n",
    "### Français\n",
    "\n",
    "Ouvrir ce notebook dans Google Colab et l’exécuter vous-même du début à la fin. Pendant que vous exécutez le notebook, cellule par cellule, pensez à ce que chaque cellule fait et essayez de faire les exercices quand vous les rencontrez.\n",
    "\n",
    "Essayez de faire les exercices mais ne vous inquiétez pas si vous n’avez pas le temps de tout faire pendant la séance appliquée et n’hésitez pas à demander aux enseignants assistants si vous avez besoin d’aide parce que vous êtes bloqués.\n",
    "\n",
    "Indice: Pour ouvrir le notebook dans Google Colab, cliquer sur l’icône de fusée au sommet de cette page puis choisissez “Colab” dans le menu déroulant."
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "colab": {
   "collapsed_sections": [],
   "provenance": [],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": "developer-developer-training-nb-maintenance-mgen-7.12.0",
   "language": "python",
   "name": "conda-env-developer-developer-training-nb-maintenance-mgen-7.12.0-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
